**GITHUB:** 

https://github.com/PUCP-DA2020-ML/Default-CreditCard

###############################################################################
# **1. COMPRENSIÓN DEL NEGOCIO**
###############################################################################

### Dataset

Usaremos el dataset de **DEFAULT CREDIT CARD CLIENTS**, el cual está disponible en:
<br/>
<br/>
https://archive.ics.uci.edu/ml/datasets/default+of+credit+card+clients
<br/>
<br/>
El objetivo será predecir si el crédito de un cliente caerá en incumplimiento de pago.

**Descripción de los datos:**

>Nro | Variable | Tipo_Variable | Descripción | Dominio
>--- | --- | --- | --- | ---
>1 | LIMIT_BAL (X1) | Numérica (Continua) | Monto del crédito otorgado (NT dollar) | 
>2 | SEX (X2) | Categórica (Nominal) | Género | (1 = masculino; 2 = femenino)
>3 | EDUCATION (X3) | Categórica (Ordinal) | Educación | (1 = escuela de posgrado; 2 = universidad; 3 = secundaria; 4 = otros).
>4 | MARRIAGE (X4) | Categórica (Nominal) | Estado Civil | (1 = casado; 2 = soltero; 3 = otros)
>5 | AGE (X5) | Numérica (Discreta) | Edad (Años) | 
>6 | PAY_0 (X6) | Categórica (Ordinal) | El estado de reembolso en Setiembre del 2005 | (-1 = pago al día; 1 = 1 mes de retraso en el pago; 2 = 2 meses de retraso en el pago; . . .; 8 = 8 meses de retraso en el pago; 9 = 9 meses o más de retraso en el pago)
>7 | PAY_2 (X7) | Categórica (Ordinal) | El estado de reembolso en Agosto del 2005 | (-1 = pago al día; 1 = 1 mes de retraso en el pago; 2 = 2 meses de retraso en el pago; . . .; 8 = 8 meses de retraso en el pago; 9 = 9 meses o más de retraso en el pago)
>8 | PAY_3 (X8) | Categórica (Ordinal) | El estado de reembolso en Julio del 2005 | (-1 = pago al día; 1 = 1 mes de retraso en el pago; 2 = 2 meses de retraso en el pago; . . .; 8 = 8 meses de retraso en el pago; 9 = 9 meses o más de retraso en el pago)
>9 | PAY_4 (X9) | Categórica (Ordinal) | El estado de reembolso en Junio del 2005 | (-1 = pago al día; 1 = 1 mes de retraso en el pago; 2 = 2 meses de retraso en el pago; . . .; 8 = 8 meses de retraso en el pago; 9 = 9 meses o más de retraso en el pago)
>10 | PAY_5 (X10) | Categórica (Ordinal) | El estado de reembolso en Mayo del 2005 | (-1 = pago al día; 1 = 1 mes de retraso en el pago; 2 = 2 meses de retraso en el pago; . . .; 8 = 8 meses de retraso en el pago; 9 = 9 meses o más de retraso en el pago)
>11 | PAY_6 (X11) | Categórica (Ordinal) | El estado de reembolso en Abril del 2005 | (-1 = pago al día; 1 = 1 mes de retraso en el pago; 2 = 2 meses de retraso en el pago; . . .; 8 = 8 meses de retraso en el pago; 9 = 9 meses o más de retraso en el pago)
>12 | BILL_AMT1 (X12) | Numérica (Continua) | Monto del estado de cuenta en Setiembre del 2005 (NT dollar) | 
>13 | BILL_AMT2 (X13) | Numérica (Continua) | Monto del estado de cuenta en Agosto del 2005 (NT dollar) | 
>14 | BILL_AMT3 (X14) | Numérica (Continua) | Monto del estado de cuenta en Julio del 2005 (NT dollar) | 
>15 | BILL_AMT4 (X15) | Numérica (Continua) | Monto del estado de cuenta en Junio del 2005 (NT dollar) | 
>16 | BILL_AMT5 (X16) | Numérica (Continua) | Monto del estado de cuenta en Mayo del 2005 (NT dollar) | 
>17 | BILL_AMT6 (X17) | Numérica (Continua) | Monto del estado de cuenta en Abril del 2005 (NT dollar) | 
>18 | PAY_AMT1 (X18) | Numérica (Continua) | Monto pagado en Setiembre del 2005 (NT dollar) | 
>19 | PAY_AMT2 (X19) | Numérica (Continua) | Monto pagado en Agosto del 2005 (NT dollar) | 
>20 | PAY_AMT3 (X20) | Numérica (Continua) | Monto pagado en Julio del 2005 (NT dollar) | 
>21 | PAY_AMT4 (X21) | Numérica (Continua) | Monto pagado en Junio del 2005 (NT dollar) | 
>22 | PAY_AMT5 (X22) | Numérica (Continua) | Monto pagado en Mayo del 2005 (NT dollar) | 
>23 | PAY_AMT6 (X23) | Numérica (Continua) | Monto pagado en Abril del 2005 (NT dollar) | 
>24 | DEFAULT_PAYMENT_NEXT_MONTH (Y) | Categórica (Nominal binaria) | Incumplimiento de pago del crédito como variable objetivo o respuesta | (Si = 1, No = 0)

###############################################################################
# **2. COMPRENSIÓN DE LOS DATOS**
###############################################################################

In [81]:
!pip install funpymodeling

### Importación de Librerías

In [82]:
# Importando librerias
import pandas as pd
#import graphviz
import numpy as np
from sklearn import tree
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()
from pandas_profiling import ProfileReport
from funpymodeling.exploratory import profiling_num, freq_tbl, status, cat_vars, num_vars
import pickle
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, roc_curve, roc_auc_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import plot_roc_curve
from sklearn.metrics import auc
from sklearn.model_selection import GridSearchCV
from sklearn import metrics
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import validation_curve
from sklearn.model_selection import learning_curve
from pprint import pprint
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB

## **IMPORTACIÓN DE DATOS** (datasets)

In [83]:
#from google.colab import files
#subido = files.upload()
#import io
#df = pd.read_excel(io.BytesIO(subido['default_of_credit_card_clients.xls']), header =1)
#df = pd.read_excel('default_of_credit_card_clients.xls', header =1)

In [84]:
#Importación del dataset
df_defcredcard = pd.read_csv('https://raw.githubusercontent.com/PUCP-DA2020-ML/Default-CreditCard/main/default-of-credit-card-clients.csv', sep=';')

In [85]:
#Tamaño del dataset (registros, variables)
df_defcredcard.shape

(30000, 25)

In [86]:
#Muestra (5 registros) del dataset
df_defcredcard.head()

ID  LIMIT_BAL  SEX  ...  PAY_AMT5  PAY_AMT6  default_payment_next_month
0   1      20000    2  ...         0         0                           1
1   2     120000    2  ...         0      2000                           1
2   3      90000    2  ...      1000      5000                           0
3   4      50000    2  ...      1069      1000                           0
4   5      50000    1  ...       689       679                           0

[5 rows x 25 columns]

In [87]:
# Eliminar columna de correlativo
df_defcredcard.drop(['ID'], axis=1, inplace=True)

## **ANÁLISIS EXPLORATORIO DE DATOS (AED)**

In [88]:
target = 'default_payment_next_month'
predictors = ['LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 
                'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 
                'BILL_AMT1','BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6',
                'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6']

In [89]:
df_defcredcard["SEX"] = df_defcredcard["SEX"].astype(str)
df_defcredcard["EDUCATION"] = df_defcredcard["EDUCATION"].astype(str)
df_defcredcard["MARRIAGE"] = df_defcredcard["MARRIAGE"].astype(str)
df_defcredcard["PAY_0"] = df_defcredcard["PAY_0"].astype(str)
df_defcredcard["PAY_2"] = df_defcredcard["PAY_2"].astype(str)
df_defcredcard["PAY_3"] = df_defcredcard["PAY_3"].astype(str)
df_defcredcard["PAY_4"] = df_defcredcard["PAY_4"].astype(str)
df_defcredcard["PAY_5"] = df_defcredcard["PAY_5"].astype(str)
df_defcredcard["PAY_6"] = df_defcredcard["PAY_6"].astype(str)
df_defcredcard["default_payment_next_month"] = df_defcredcard["default_payment_next_month"].astype(str)

### ***Resumen Descriptivo***

In [90]:
#------------------------------------------------
# ANÁLISIS DE EXPLORACIÓN RÁPIDA O INICIAL
#------------------------------------------------
#from pandas_profiling import ProfileReport
#ProfileReport(df_defcredcard, minimal=True)

In [91]:
#------------------------------------------------
# TIPOS DE DATOS
#------------------------------------------------
df_defcredcard.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   LIMIT_BAL                   30000 non-null  int64 
 1   SEX                         30000 non-null  object
 2   EDUCATION                   30000 non-null  object
 3   MARRIAGE                    30000 non-null  object
 4   AGE                         30000 non-null  int64 
 5   PAY_0                       30000 non-null  object
 6   PAY_2                       30000 non-null  object
 7   PAY_3                       30000 non-null  object
 8   PAY_4                       30000 non-null  object
 9   PAY_5                       30000 non-null  object
 10  PAY_6                       30000 non-null  object
 11  BILL_AMT1                   30000 non-null  int64 
 12  BILL_AMT2                   30000 non-null  int64 
 13  BILL_AMT3                   30000 non-null  in

In [92]:
#--------------------------------------------------------------
# RESUMEN DESCRIPTIVO
#--------------------------------------------------------------
#df_defcredcard.describe(include='all').T
profiling_num(df_defcredcard)

variable           mean        std_dev  ...     p_0.75     p_0.95     p_0.99
0   LIMIT_BAL  167484.322667  129747.661567  ...  240000.00  430000.00  500000.00
1         AGE      35.485500       9.217904  ...      41.00      53.00      60.00
2   BILL_AMT1   51223.330900   73635.860576  ...   67091.00  201203.05  350110.68
3   BILL_AMT2   49179.075167   71173.768783  ...   64006.25  194792.20  337495.28
4   BILL_AMT3   47013.154800   69349.387427  ...   60164.75  187821.05  325030.39
5   BILL_AMT4   43262.948967   64332.856134  ...   54506.00  174333.35  304997.27
6   BILL_AMT5   40311.400967   60797.155770  ...   50190.50  165794.30  285868.33
7   BILL_AMT6   38871.760400   59554.107537  ...   49198.25  161912.00  279505.06
8    PAY_AMT1    5663.580500   16563.280354  ...    5006.00   18428.20   66522.18
9    PAY_AMT2    5921.163500   23040.870402  ...    5000.00   19004.35   76651.02
10   PAY_AMT3    5225.681500   17606.961470  ...    4505.00   17589.40   70000.00
11   PAY_AMT4    4826.076867   15666.159744  ...    4013.25   16014.95   67054.44
12   PAY_AMT5    4799.387633   15278.305679  ...    4031.50   16000.00   65607.56
13   PAY_AMT6    5215.502567   17777.465775  ...    4000.00   17343.80   82619.05

[14 rows x 11 columns]

Comentario:
- Hay variables con rangos mas grandes que otras, por lo que se debe escalar los valores para que el grado en el que afectarán a la variable objetivo podría ser igual y evitar sesgos hacia variables con un rango mayor.

#### Revisión de Valores Únicos y Valores Perdidos (Missings Values)

In [93]:
#--------------------------------------------------------------------------------------------------
# TABLA DE ESTADO DE LAS VARIABLES (Valores Nulos / Valores Únicos / Valores Perdidos / Ceros )
#--------------------------------------------------------------------------------------------------
status(df_defcredcard)

variable  q_nan  p_nan  q_zeros   p_zeros  unique    type
0                    LIMIT_BAL      0    0.0        0  0.000000      81   int64
1                          SEX      0    0.0        0  0.000000       2  object
2                    EDUCATION      0    0.0        0  0.000000       7  object
3                     MARRIAGE      0    0.0        0  0.000000       4  object
4                          AGE      0    0.0        0  0.000000      56   int64
5                        PAY_0      0    0.0        0  0.000000      11  object
6                        PAY_2      0    0.0        0  0.000000      11  object
7                        PAY_3      0    0.0        0  0.000000      11  object
8                        PAY_4      0    0.0        0  0.000000      11  object
9                        PAY_5      0    0.0        0  0.000000      10  object
10                       PAY_6      0    0.0        0  0.000000      10  object
11                   BILL_AMT1      0    0.0     2008  0.066933   22723   int64
12                   BILL_AMT2      0    0.0     2506  0.083533   22346   int64
13                   BILL_AMT3      0    0.0     2870  0.095667   22026   int64
14                   BILL_AMT4      0    0.0     3195  0.106500   21548   int64
15                   BILL_AMT5      0    0.0     3506  0.116867   21010   int64
16                   BILL_AMT6      0    0.0     4020  0.134000   20604   int64
17                    PAY_AMT1      0    0.0     5249  0.174967    7943   int64
18                    PAY_AMT2      0    0.0     5396  0.179867    7899   int64
19                    PAY_AMT3      0    0.0     5968  0.198933    7518   int64
20                    PAY_AMT4      0    0.0     6408  0.213600    6937   int64
21                    PAY_AMT5      0    0.0     6703  0.223433    6897   int64
22                    PAY_AMT6      0    0.0     7173  0.239100    6939   int64
23  default_payment_next_month      0    0.0        0  0.000000       2  object

##### Valores Únicos

In [94]:
#--------------------------------------------------
# Cantidad de Valores Únicos
#--------------------------------------------------
# df_defcredcard.nunique()

In [95]:
#--------------------------------------------------
# Valores Únicos por Columnas
#--------------------------------------------------
for columna in df_defcredcard.columns:
    print(f'Valores únicos de la columna {columna}: {(df_defcredcard[columna].unique())}')

Valores únicos de la columna LIMIT_BAL: [  20000  120000   90000   50000  500000  100000  140000  200000  260000
  630000   70000  250000  320000  360000  180000  130000  450000   60000
  230000  160000  280000   10000   40000  210000  150000  380000  310000
  400000   80000  290000  340000  300000   30000  240000  470000  480000
  350000  330000  110000  420000  170000  370000  270000  220000  190000
  510000  460000  440000  410000  490000  390000  580000  600000  620000
  610000  700000  670000  680000  430000  550000  540000 1000000  530000
  710000  560000  520000  750000  640000   16000  570000  590000  660000
  720000  327680  740000  800000  760000  690000  650000  780000  730000]
Valores únicos de la columna SEX: ['2' '1']
Valores únicos de la columna EDUCATION: ['2' '1' '3' '5' '4' '6' '0']
Valores únicos de la columna MARRIAGE: ['1' '2' '3' '0']
Valores únicos de la columna AGE: [24 26 34 37 57 29 23 28 35 51 41 30 49 39 40 27 47 33 32 54 58 22 25 31
 46 42 43 45 56 44 53 38

Comentario: 
- La variable EDUCATION tienen categorías adicionales a lo definido en el diccionario (0, 5, 6)
- La variable MARRIAGE tiene una categoría adicional a lo definido en el diccionario (0)
- Las variables de historial de pago pasado (PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6) tienen valores adicionales a lo definido en el el diccionario ('0' y'-2'), que son ambiguos. Quizás '0' y '-2' se refieren también al pago al día o incluso pago antes de la fecha límite.

In [96]:
#--------------------------------------------------
# % de Valores Únicos
#--------------------------------------------------
(df_defcredcard.nunique()/len(df_defcredcard))*100

LIMIT_BAL                      0.270000
SEX                            0.006667
EDUCATION                      0.023333
MARRIAGE                       0.013333
AGE                            0.186667
PAY_0                          0.036667
PAY_2                          0.036667
PAY_3                          0.036667
PAY_4                          0.036667
PAY_5                          0.033333
PAY_6                          0.033333
BILL_AMT1                     75.743333
BILL_AMT2                     74.486667
BILL_AMT3                     73.420000
BILL_AMT4                     71.826667
BILL_AMT5                     70.033333
BILL_AMT6                     68.680000
PAY_AMT1                      26.476667
PAY_AMT2                      26.330000
PAY_AMT3                      25.060000
PAY_AMT4                      23.123333
PAY_AMT5                      22.990000
PAY_AMT6                      23.130000
default_payment_next_month     0.006667
dtype: float64

Comentario:

No hay variables que tengan un alto porcentaje de valores únicos (referencia: 70-80%)

In [97]:
por_val_unicos = pd.DataFrame(data = df_defcredcard.nunique()/len(df_defcredcard), columns=['Por_Val_Unico'])
por_val_unicos[por_val_unicos['Por_Val_Unico'] > 0.80].count()

Por_Val_Unico    0
dtype: int64

##### Valores Perdidos (Missings Values)

In [98]:
#--------------------------------------------
# Cantidad de Valores Nulos
#--------------------------------------------
# df_defcredcard.isna().sum()

In [99]:
#--------------------------------------------
# % de Valores Nulos
#--------------------------------------------
# (df_defcredcard.isna().sum()/len(df_defcredcard))*100

Comentario:
- No hay registros (filas) ni variables (columnas) que tengan valores nulos

### ***Análisis Univariado***

In [100]:
#--------------------------------------------
# TIPO DE VARIABLES
#--------------------------------------------
#from funpymodeling.exploratory import cat_vars, num_vars
print("Variables Categóricas: ", cat_vars(df_defcredcard))
print("Variables Numéricas: ", num_vars(df_defcredcard))

Variables Categóricas:  Index(['SEX', 'EDUCATION', 'MARRIAGE', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4',
       'PAY_5', 'PAY_6', 'default_payment_next_month'],
      dtype='object')
Variables Numéricas:  Index(['LIMIT_BAL', 'AGE', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4',
       'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3',
       'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6'],
      dtype='object')


#### Análisis de Variables Categóricas

In [101]:
#--------------------------------------------
# TABLAS DE FRECUENCIAS
#--------------------------------------------
freq_tbl(df_defcredcard)
#freq_tbl(df_defcredcard['default_payment_next_month'])

  SEX  frequency  percentage  cumulative_perc
0   2      18112    0.603733         0.603733
1   1      11888    0.396267         1.000000

----------------------------------------------------------------

  EDUCATION  frequency  percentage  cumulative_perc
0         2      14030    0.467667         0.467667
1         1      10585    0.352833         0.820500
2         3       4917    0.163900         0.984400
3         5        280    0.009333         0.993733
4         4        123    0.004100         0.997833
5         6         51    0.001700         0.999533
6         0         14    0.000467         1.000000

----------------------------------------------------------------

  MARRIAGE  frequency  percentage  cumulative_perc
0        2      15964    0.532133         0.532133
1        1      13659    0.455300         0.987433
2        3        323    0.010767         0.998200
3        0         54    0.001800         1.000000

--------------------------------------------------------

In [102]:
#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable SEX
#---------------------------------------------------------------------------------
#pd.DataFrame({'Cantidad': df_defcredcard['SEX'].value_counts(), '%': df_defcredcard['SEX'].value_counts(normalize=True) * 100})

#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable EDUCATION
#---------------------------------------------------------------------------------
# pd.DataFrame({'Cantidad': df_defcredcard['EDUCATION'].value_counts(), '%': df_defcredcard['EDUCATION'].value_counts(normalize=True) * 100})

#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable MARRIAGE
#---------------------------------------------------------------------------------
# pd.DataFrame({'Cantidad': df_defcredcard['MARRIAGE'].value_counts(), '%': df_defcredcard['MARRIAGE'].value_counts(normalize=True) * 100})

#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable PAY_0
#---------------------------------------------------------------------------------
# pd.DataFrame({'Cantidad': df_defcredcard['PAY_0'].value_counts(), '%': df_defcredcard['PAY_0'].value_counts(normalize=True) * 100})

#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable PAY_2
#---------------------------------------------------------------------------------
# pd.DataFrame({'Cantidad': df_defcredcard['PAY_2'].value_counts(), '%': df_defcredcard['PAY_2'].value_counts(normalize=True) * 100})

#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable PAY_3
#---------------------------------------------------------------------------------
# pd.DataFrame({'Cantidad': df_defcredcard['PAY_3'].value_counts(), '%': df_defcredcard['PAY_3'].value_counts(normalize=True) * 100})

#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable PAY_4
#---------------------------------------------------------------------------------
# pd.DataFrame({'Cantidad': df_defcredcard['PAY_4'].value_counts(), '%': df_defcredcard['PAY_4'].value_counts(normalize=True) * 100})

#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable PAY_5
#---------------------------------------------------------------------------------
# pd.DataFrame({'Cantidad': df_defcredcard['PAY_5'].value_counts(), '%': df_defcredcard['PAY_5'].value_counts(normalize=True) * 100})

#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable PAY_6
#---------------------------------------------------------------------------------
# pd.DataFrame({'Cantidad': df_defcredcard['PAY_6'].value_counts(), '%': df_defcredcard['PAY_6'].value_counts(normalize=True) * 100})

In [103]:
#--------------------------------------------------
# GRÁFICOS DE BARRA
#--------------------------------------------------
# Esta función estara en funpymodeling como freq() o plot_freq()
# NOTA: antes de usar este script, saquen las variables que tienen mas de 100 categorias distintas
import matplotlib.pyplot as plt
import seaborn as sns

d_plot = df_defcredcard
fig, ax = plt.subplots(5, 2, figsize=(20, 20))
for variable, subplot in zip(cat_vars(d_plot), ax.flatten()):
    sns.countplot(y=d_plot[variable], ax=subplot, order = d_plot[variable].value_counts().index)
    for label in subplot.get_xticklabels():
        label.set_rotation(90)

In [104]:
#plt.figure(figsize = (15, 15))
#plt.subplot(3, 3, 1)
#sns.countplot(data=df_defcredcard, x='SEX')
#plt.subplot(3, 3, 2)
#sns.countplot(data=df_defcredcard, x='EDUCATION')
#plt.subplot(3, 3, 3)
#sns.countplot(data=df_defcredcard, x='MARRIAGE')
#plt.subplot(3, 3, 4)
#sns.countplot(data=df_defcredcard, x='PAY_0')
#plt.subplot(3, 3, 5)
#sns.countplot(data=df_defcredcard, x='PAY_2')
#plt.subplot(3, 3, 6)
#sns.countplot(data=df_defcredcard, x='PAY_3')
#plt.subplot(3, 3, 7)
#sns.countplot(data=df_defcredcard, x='PAY_4')
#plt.subplot(3, 3, 8)
#sns.countplot(data=df_defcredcard, x='PAY_5')
#plt.subplot(3, 3, 9)
#sns.countplot(data=df_defcredcard, x='PAY_6')
#plt.tight_layout()
#plt.show()

Comentario:
- La variable objetivo default_payment_next_month no tiene una proporción balanceada de sus clases:

0 : 77.88 %

1 : 22.12 %

#### Análisis de Variables Numéricas

In [105]:
#----------------------------------------------
# HISTOGRAMAS
#----------------------------------------------
df_defcredcard.hist(figsize = (20,20))

array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7fb32119e6a0>,
      dtype=object)

In [106]:
#---------------------------------------------
# DIAGRAMAS DE CAJA
#---------------------------------------------
#df.boxplot(figsize=(20,10))

# create figure and axes
fig, axes = plt.subplots(5,3, figsize=(20, 20))

for i, el in enumerate(list(df_defcredcard.select_dtypes(exclude='object').columns.values)):
    a = df_defcredcard.boxplot(el, ax=axes.flatten()[i])

plt.tight_layout() 
plt.show()

In [107]:
#sns.pairplot(df_defcredcard, hue="default_payment_next_month", height=5, diag_kind="hist")

### ***Análisis Bivariado***

In [108]:
#from funpymodeling.exploratory import cat_vars, num_vars
print("Variables Categóricas: ", cat_vars(df_defcredcard))

Variables Categóricas:  Index(['SEX', 'EDUCATION', 'MARRIAGE', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4',
       'PAY_5', 'PAY_6', 'default_payment_next_month'],
      dtype='object')


#### Análisis de Var. Categórica vs Categórica

In [109]:
#---------------------------------------------------------------
# TABLAS DE CONTINGENCIA BIVARIADA
#---------------------------------------------------------------

#...............................................................
# Análisis: INCUMPLIMIENTO DE PAGO vs GÉNERO
#...............................................................
# Porcentages totales (suma de celdas 100%)
#pd.crosstab(df_defcredcard['SEX'], df_defcredcard['default_payment_next_month'], margins=True, normalize=True)

#df_target_vs_sex = df_defcredcard.groupby(['SEX',target])[target].count().unstack()
#df_target_vs_sex['percentage'] = round((df_target_vs_sex[1]/(df_target_vs_sex[0]+df_target_vs_sex[1]))*100,2)
#df_target_vs_sex

# Cantidades Totales (suma de celdas 100%) 
#pd.crosstab(df_defcredcard['SEX'], df_defcredcard['default_payment_next_month'], margins=True)

# Porcentages por registro (útil para selección de variables)
# pd.crosstab(df_defcredcard['SEX'], df_defcredcard['default_payment_next_month'], normalize='index', margins=True)

# Cantidades por registro 
pd.crosstab(df_defcredcard['SEX'], df_defcredcard['default_payment_next_month'], margins='index')

default_payment_next_month      0     1    All
SEX                                           
1                            9015  2873  11888
2                           14349  3763  18112
All                         23364  6636  30000

In [110]:
#...............................................................
# Análisis: INCUMPLIMIENTO DE PAGO vs EDUCACIÓN
#...............................................................
# Porcentages totales (suma de celdas 100%)
#pd.crosstab(df_defcredcard['EDUCATION'], df_defcredcard['default_payment_next_month'], margins=True, normalize=True)

#df_target_vs_educ = df_defcredcard.groupby(['EDUCATION',target])[target].count().unstack()
#df_target_vs_educ['percentage'] = round((df_target_vs_educ[1]/(df_target_vs_educ[0]+df_target_vs_educ[1]))*100,2)
#df_target_vs_educ

# Cantidades Totales (suma de celdas 100%) 
#pd.crosstab(df_defcredcard['EDUCATION'], df_defcredcard['default_payment_next_month'], margins=True)

# Porcentages por registro (útil para selección de variables)
# pd.crosstab(df_defcredcard['EDUCATION'], df_defcredcard['default_payment_next_month'], normalize='index', margins=True)

# Cantidades por registro 
pd.crosstab(df_defcredcard['EDUCATION'], df_defcredcard['default_payment_next_month'], margins='index')

default_payment_next_month      0     1    All
EDUCATION                                     
0                              14     0     14
1                            8549  2036  10585
2                           10700  3330  14030
3                            3680  1237   4917
4                             116     7    123
5                             262    18    280
6                              43     8     51
All                         23364  6636  30000

In [111]:
#...............................................................
# Análisis: INCUMPLIMIENTO DE PAGO vs ESTADO CIVIL
#...............................................................
# Porcentages totales (suma de celdas 100%)
#pd.crosstab(df_defcredcard['MARRIAGE'], df_defcredcard['default_payment_next_month'], margins=True, normalize=True)

#df_target_vs_marriage = df_defcredcard.groupby(['MARRIAGE',target])[target].count().unstack()
#df_target_vs_marriage['percentage'] = round((df_target_vs_marriage[1]/(df_target_vs_marriage[0]+df_target_vs_marriage[1]))*100,2)
#df_target_vs_marriage

# Cantidades Totales (suma de celdas 100%) 
#pd.crosstab(df_defcredcard['MARRIAGE'], df_defcredcard['default_payment_next_month'], margins=True)

# Porcentages por registro (útil para selección de variables)
# pd.crosstab(df_defcredcard['MARRIAGE'], df_defcredcard['default_payment_next_month'], normalize='index', margins=True)

# Cantidades por registro 
pd.crosstab(df_defcredcard['MARRIAGE'], df_defcredcard['default_payment_next_month'], margins='index')

default_payment_next_month      0     1    All
MARRIAGE                                      
0                              49     5     54
1                           10453  3206  13659
2                           12623  3341  15964
3                             239    84    323
All                         23364  6636  30000

In [112]:
#---------------------------------------------------------------
# DIAGRAMAS DE BARRA BIVARIADO
#---------------------------------------------------------------

fig,axes = plt.subplots(5,2,figsize=(15,20))
for idx, cat_col in enumerate(cat_vars(df_defcredcard)):
    row,col = idx//2,idx%2
    sns.countplot(x=cat_col,data=df_defcredcard,hue='default_payment_next_month',ax=axes[row,col])

#g = sns.catplot(x="AGE", col="default_payment_next_month", data = df_defcredcard, kind="count")

Comentario: 
- Los hombres tienen una mayor probabilidad de incumplir que las mujeres, ya que ~ 24% de los hombres incumplen en comparación con ~ 20% de las mujeres. Además, según los datos, un mayor número de mujeres han incumplido (3763) que de hombres (2873) en los meses anteriores.
- Es visible que una mayor cantidad de educación tiende a reducir la probabilidad de incumplimiento. Esto tiene sentido porque una educación más baja puede significar un salario y un límite más bajos.
- Los clientes casados ​​tienen una mayor probabilidad de incumplimiento que los clientes solteros


#### Análisis de Var. Numérica vs Categórica

In [113]:
#df_defcredcard.groupby('MARRIAGE')['AGE'].mean().sort_values(ascending=False)
#df_defcredcard.groupby('MARRIAGE')['AGE'].describe()
#df_defcredcard.groupby('default_payment_next_month').mean()

In [114]:
#sns.boxplot(y = 'AGE', x = 'MARRIAGE', data = df_defcredcard)
#pd.DataFrame(df_defcredcard.groupby('Purpose')['AGE'].median().sort_values(ascending=True))

#grouped = df_defcredcard.loc[:,['AGE', 'MARRIAGE']].groupby(['MARRIAGE']).median().sort_values(by='AGE')
#grouped['MARRIAGE'] = grouped.index
#grouped.head()

#fig, ax = plt.subplots(figsize=(8,6))
#bp = sns.boxplot(y = 'AGE', x = 'MARRIAGE', data = df_defcredcard, order = grouped.index)
#bp.set_xticklabels(bp.get_xmajorticklabels(), fontsize = 10, rotation=45)
#plt.show()

In [115]:
#sns.violinplot(y='AGE', x='MARRIAGE', data = df_defcredcard)

#### Análisis de Var. Numérica vs Numérica

In [116]:
#-------------------------------------
# MATRIZ DE CORRELACIÓN
#-------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt
plt.figure(figsize = (30,10)) # tamaño de la imagen
corr = df_defcredcard.iloc[:,0:].corr()

sns.heatmap(corr, annot=True, fmt='.4g',linewidths=0.5,
            xticklabels=corr.columns.values,
            yticklabels=corr.columns.values)

Comentario:

- Ninguna de las variables predictora tiene una fuerte correlación con la variable objetivo. 
- La variable 'PAY_6' tienen una pequeña cantidad de correlación positiva con la variable objetivo y una correlación más fuerte entre las variables de historial de pago. Entonces, si ha incumplido o no en los meses anteriores, podría decidir la probabilidad de incumplimiento en el próximo mes. 
- Las variables de monto de estado de cuenta tienen una fuerte correlación positiva entre ellas. Esto también tiene sentido ya que los patrones de gasto de los clientes tienden a permanecer constantes y las organizaciones de comercio electrónico se aprovechan de esto. 
- La correlación negativa más fuerte se da entre 'AGE' y 'MARRIAGE', que se espera ya que los clientes más jóvenes tienden a ser solteros y los mayores tienden a estar casados.

In [117]:
#-----------------------------------------------
# DIAGRAMA SCATTER PLOT
#-----------------------------------------------
sns.pairplot(df_defcredcard)

###############################################################################
# **3. PREPARACIÓN DE LOS DATOS**
###############################################################################

## **PREPARACIÓN DE LOS DATOS (ENTRENAMIENTO)**

#### **PRE-PROCESAMIENTO DE DATOS**

In [118]:
df_defcredcard_preproc = df_defcredcard.copy()

In [119]:
df_defcredcard.shape

(30000, 24)

In [120]:
df_defcredcard.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   LIMIT_BAL                   30000 non-null  int64 
 1   SEX                         30000 non-null  object
 2   EDUCATION                   30000 non-null  object
 3   MARRIAGE                    30000 non-null  object
 4   AGE                         30000 non-null  int64 
 5   PAY_0                       30000 non-null  object
 6   PAY_2                       30000 non-null  object
 7   PAY_3                       30000 non-null  object
 8   PAY_4                       30000 non-null  object
 9   PAY_5                       30000 non-null  object
 10  PAY_6                       30000 non-null  object
 11  BILL_AMT1                   30000 non-null  int64 
 12  BILL_AMT2                   30000 non-null  int64 
 13  BILL_AMT3                   30000 non-null  in

#### Eliminación de Registros Duplicados

In [121]:
df_defcredcard_preproc.drop_duplicates(keep="first", inplace=True)

In [122]:
df_defcredcard_preproc.shape

(29965, 24)

#### Eliminación de Valores Únicos

In [123]:
#Comentario:
#No se encontró registros con % altos de valores únicos para eliminación
pd.DataFrame(df_defcredcard_preproc.nunique()/len(df_defcredcard_preproc))[pd.DataFrame(df_defcredcard_preproc.nunique()/len(df_defcredcard_preproc))[0] > 0.80]

Empty DataFrame
Columns: [0]
Index: []

In [124]:
df_defcredcard_preproc.shape

(29965, 24)

#### Eliminación de Variables Correlacionadas

In [125]:
# Comentario:
# No se eliminará variables correlacionadas
# Las variables de historial de estado de cuenta tienen un grado de correlación fuerte, 
# sin embargo, cada una de estas variables mide la cantidad de meses que no se paga

#### Conversión de Variables de Formato Texto a Numérico / División de Variables Numéricas y Categóricas

In [126]:
#Comentario:
#Todas las variables son numéricas, no se requiere conversión ni división

In [127]:
# numeric_columns = list()
# categorical_columns = list()
# columns_with_multiple_values = set()
# dictionary_of_columns_with_index_to_categorical = dict()
# dictionary_of_columns_with_categorical_to_index = dict()

# for column in df_defcredcard_preproc:
#   #Determinamos si la variable es numérica o no
#   if pd.api.types.is_numeric_dtype(df_defcredcard_preproc[column]):
#     numeric_columns.append(column)
#   else:
#     features = set()

#     for row in df_defcredcard_preproc[column]:
#       if isinstance(row, str):
#         if (column not in columns_with_multiple_values) and (";" in row):
#           columns_with_multiple_values.add(column)
        
#         values = row.split(";")
#       else:
#         continue
      
#       for value in values:
#         if value not in features:
#           features.add(value)
    
#     dictionary_of_columns_with_index_to_categorical[column] = dict()
#     dictionary_of_columns_with_categorical_to_index[column] = dict()

#     #Indexamos los valores (categorías), sin tomar en consideración los nulos, de la variable y guardamos esa información en los diccionarios
#     for index, category in enumerate(features):
#       dictionary_of_columns_with_index_to_categorical[column][index] = category
#       dictionary_of_columns_with_categorical_to_index[column][category] = index
    
#     categorical_columns.append(column)

In [128]:
print('Variables Categoricas: ', cat_vars(df_defcredcard_preproc).to_list())

Variables Categoricas:  ['SEX', 'EDUCATION', 'MARRIAGE', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'default_payment_next_month']


In [129]:
print('Categorias de variable SEX: ', set(df_defcredcard_preproc['SEX']))
print('Categorias de variable EDUCATION: ', set(df_defcredcard_preproc['EDUCATION']))
print('Categorias de variable MARRIAGE: ', set(df_defcredcard_preproc['MARRIAGE']))
print('Categorias de variable PAY_0: ', set(df_defcredcard_preproc['PAY_0']))
print('Categorias de variable PAY_2: ', set(df_defcredcard_preproc['PAY_2']))
print('Categorias de variable PAY_3: ', set(df_defcredcard_preproc['PAY_3']))
print('Categorias de variable PAY_4: ', set(df_defcredcard_preproc['PAY_4']))
print('Categorias de variable PAY_5: ', set(df_defcredcard_preproc['PAY_5']))
print('Categorias de variable PAY_6: ', set(df_defcredcard_preproc['PAY_6']))
print('Categorias de variable default_payment_next_month: ', set(df_defcredcard_preproc['default_payment_next_month']))

Categorias de variable SEX:  {'1', '2'}
Categorias de variable EDUCATION:  {'0', '1', '5', '3', '2', '4', '6'}
Categorias de variable MARRIAGE:  {'3', '1', '2', '0'}
Categorias de variable PAY_0:  {'0', '1', '5', '-1', '6', '8', '2', '3', '-2', '4', '7'}
Categorias de variable PAY_2:  {'0', '1', '5', '-1', '6', '8', '3', '2', '-2', '4', '7'}
Categorias de variable PAY_3:  {'0', '1', '5', '-1', '8', '3', '2', '-2', '4', '7', '6'}
Categorias de variable PAY_4:  {'0', '1', '5', '-1', '6', '8', '3', '2', '-2', '4', '7'}
Categorias de variable PAY_5:  {'0', '5', '-1', '6', '8', '3', '2', '-2', '4', '7'}
Categorias de variable PAY_6:  {'0', '5', '-1', '8', '3', '2', '-2', '4', '7', '6'}
Categorias de variable default_payment_next_month:  {'0', '1'}


In [130]:
# Nota:
# No se aplicará este tipo de conversiones de formato texto a formato numérico, 
# en su lugar, se aplicará la conversión One Hot Encoding (que convierte una var. categórica a numérica y a su vez convierte texto a número: 0 o 1)
#
# df_defcredcard_preproc['SEX'].replace(to_replace=['1', '2'], value=[1, 2], inplace=True)
# df_defcredcard_preproc['EDUCATION'].replace(to_replace=['1', '2', '3', '4', '5', '6', '0'], value=[1, 2, 3, 4, 5, 6, 0], inplace=True)
# df_defcredcard_preproc['MARRIAGE'].replace(to_replace=['1', '2', '3', '0'], value=[1, 2, 3, 0], inplace=True)
# df_defcredcard_preproc['PAY_0'].replace(to_replace=['-1', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-2', '0'], value=[-1, 1, 2, 3, 4, 5, 6, 7, 8, 9, -2, 0], inplace=True)
# df_defcredcard_preproc['PAY_2'].replace(to_replace=['-1', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-2', '0'], value=[-1, 1, 2, 3, 4, 5, 6, 7, 8, 9, -2, 0], inplace=True)
# df_defcredcard_preproc['PAY_3'].replace(to_replace=['-1', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-2', '0'], value=[-1, 1, 2, 3, 4, 5, 6, 7, 8, 9, -2, 0], inplace=True)
# df_defcredcard_preproc['PAY_4'].replace(to_replace=['-1', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-2', '0'], value=[-1, 1, 2, 3, 4, 5, 6, 7, 8, 9, -2, 0], inplace=True)
# df_defcredcard_preproc['PAY_5'].replace(to_replace=['-1', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-2', '0'], value=[-1, 1, 2, 3, 4, 5, 6, 7, 8, 9, -2, 0], inplace=True)
# df_defcredcard_preproc['PAY_6'].replace(to_replace=['-1', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-2', '0'], value=[-1, 1, 2, 3, 4, 5, 6, 7, 8, 9, -2, 0], inplace=True)
# df_defcredcard_preproc['default_payment_next_month'].replace(to_replace=['0', '1'], value=[0, 1], inplace=True)

#### Tratamiento de Valores Perdidos

In [131]:
#--------------------------------------------------------------------
# TRATAMIENTO DE VALORES PERDIDOS PARA VAR. CATEGORICAS
#--------------------------------------------------------------------
df_defcredcard_preproc['EDUCATION'].replace(to_replace=['5', '6', '0'], value=['4', '4', '4'], inplace=True)
df_defcredcard_preproc['MARRIAGE'].replace(to_replace=['0'], value=['3'], inplace=True)

#'-999': Indeterminado
df_defcredcard_preproc['PAY_0'].replace(to_replace=['-2', '0'], value=['-999', '-999'], inplace=True)
df_defcredcard_preproc['PAY_2'].replace(to_replace=['-2', '0'], value=['-999', '-999'], inplace=True)
df_defcredcard_preproc['PAY_3'].replace(to_replace=['-2', '0'], value=['-999', '-999'], inplace=True)
df_defcredcard_preproc['PAY_4'].replace(to_replace=['-2', '0'], value=['-999', '-999'], inplace=True)
df_defcredcard_preproc['PAY_5'].replace(to_replace=['-2', '0'], value=['-999', '-999'], inplace=True)
df_defcredcard_preproc['PAY_6'].replace(to_replace=['-2', '0'], value=['-999', '-999'], inplace=True)

In [132]:
print('Categorias de variable SEX: ', set(df_defcredcard_preproc['SEX']))
print('Categorias de variable EDUCATION: ', set(df_defcredcard_preproc['EDUCATION']))
print('Categorias de variable MARRIAGE: ', set(df_defcredcard_preproc['MARRIAGE']))
print('Categorias de variable PAY_0: ', set(df_defcredcard_preproc['PAY_0']))
print('Categorias de variable PAY_2: ', set(df_defcredcard_preproc['PAY_2']))
print('Categorias de variable PAY_3: ', set(df_defcredcard_preproc['PAY_3']))
print('Categorias de variable PAY_4: ', set(df_defcredcard_preproc['PAY_4']))
print('Categorias de variable PAY_5: ', set(df_defcredcard_preproc['PAY_5']))
print('Categorias de variable PAY_6: ', set(df_defcredcard_preproc['PAY_6']))
print('Categorias de variable default_payment_next_month: ', set(df_defcredcard_preproc['default_payment_next_month']))

Categorias de variable SEX:  {'1', '2'}
Categorias de variable EDUCATION:  {'4', '3', '1', '2'}
Categorias de variable MARRIAGE:  {'3', '1', '2'}
Categorias de variable PAY_0:  {'1', '5', '-999', '-1', '6', '8', '3', '2', '4', '7'}
Categorias de variable PAY_2:  {'1', '5', '-999', '-1', '6', '8', '3', '2', '4', '7'}
Categorias de variable PAY_3:  {'1', '5', '-999', '-1', '8', '2', '3', '4', '7', '6'}
Categorias de variable PAY_4:  {'1', '5', '-999', '-1', '6', '8', '2', '3', '4', '7'}
Categorias de variable PAY_5:  {'5', '-999', '-1', '6', '8', '2', '3', '4', '7'}
Categorias de variable PAY_6:  {'5', '-999', '-1', '8', '2', '3', '4', '7', '6'}
Categorias de variable default_payment_next_month:  {'0', '1'}


In [133]:
df_defcredcard_preproc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29965 entries, 0 to 29999
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   LIMIT_BAL                   29965 non-null  int64 
 1   SEX                         29965 non-null  object
 2   EDUCATION                   29965 non-null  object
 3   MARRIAGE                    29965 non-null  object
 4   AGE                         29965 non-null  int64 
 5   PAY_0                       29965 non-null  object
 6   PAY_2                       29965 non-null  object
 7   PAY_3                       29965 non-null  object
 8   PAY_4                       29965 non-null  object
 9   PAY_5                       29965 non-null  object
 10  PAY_6                       29965 non-null  object
 11  BILL_AMT1                   29965 non-null  int64 
 12  BILL_AMT2                   29965 non-null  int64 
 13  BILL_AMT3                   29965 non-null  in

In [134]:
#--------------------------------------------------------------------
# TRATAMIENTO DE VALORES PERDIDOS PARA VAR. NUMÉRICAS
#--------------------------------------------------------------------

#### Tratamiento de Valores Nulos

In [135]:
#Comentario:
#No se encontró registros con valores nulos
pd.DataFrame(df_defcredcard_preproc.isna().sum()/len(df_defcredcard_preproc))[pd.DataFrame(df_defcredcard_preproc.isna().sum()/len(df_defcredcard_preproc))[0] > 0]

Empty DataFrame
Columns: [0]
Index: []

##### Eliminación de Valores Nulos

In [136]:
#Comentario:
#No se encontró registros con valores nulos para eliminación:
#1) Filas con Bajo % de Valores Nulos
#2) Columnas con Alto % de Valores Nulos

##### Imputación (Estimación) de Valores Nulos

In [137]:
#Comentario:
#No se encontró registros con valores nulos para imputación

In [138]:
#data['df_defcredcard_preproc'] = data['df_defcredcard_preproc'].fillna(value="nulo")

#### **CREACIÓN / TRANSFORMACION DE CARACTERÍSTICAS**

In [139]:
#Comentario:
#No se creará nuevas variables

In [140]:
df_defcredcard_preproc_OneHotEnc = pd.get_dummies(
                                                  df_defcredcard_preproc, 
                                                  drop_first=True, # Eliminar la 1ra categoria para obtener k-1 categorias (para evitar correlación de variables redundantes)
                                                  #dummy_na=True # Adicionar una columna para indicar NaN
                                                  )
# Renombrar variable objetivo
df_defcredcard_preproc_OneHotEnc.rename(columns={'default_payment_next_month_1': 'default_payment_next_month'}, inplace=True)

In [141]:
df_defcredcard_preproc_OneHotEnc.head()

LIMIT_BAL  AGE  BILL_AMT1  ...  PAY_6_7  PAY_6_8  default_payment_next_month
0      20000   24       3913  ...        0        0                           1
1     120000   26       2682  ...        0        0                           1
2      90000   34      29239  ...        0        0                           0
3      50000   37      46990  ...        0        0                           0
4      50000   57       8617  ...        0        0                           0

[5 rows x 73 columns]

In [142]:
df_defcredcard_preproc_OneHotEnc.shape

(29965, 73)

In [143]:
df_defcredcard_preproc_OneHotEnc.columns

Index(['LIMIT_BAL', 'AGE', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4',
       'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3',
       'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6', 'SEX_2', 'EDUCATION_2',
       'EDUCATION_3', 'EDUCATION_4', 'MARRIAGE_2', 'MARRIAGE_3', 'PAY_0_-999',
       'PAY_0_1', 'PAY_0_2', 'PAY_0_3', 'PAY_0_4', 'PAY_0_5', 'PAY_0_6',
       'PAY_0_7', 'PAY_0_8', 'PAY_2_-999', 'PAY_2_1', 'PAY_2_2', 'PAY_2_3',
       'PAY_2_4', 'PAY_2_5', 'PAY_2_6', 'PAY_2_7', 'PAY_2_8', 'PAY_3_-999',
       'PAY_3_1', 'PAY_3_2', 'PAY_3_3', 'PAY_3_4', 'PAY_3_5', 'PAY_3_6',
       'PAY_3_7', 'PAY_3_8', 'PAY_4_-999', 'PAY_4_1', 'PAY_4_2', 'PAY_4_3',
       'PAY_4_4', 'PAY_4_5', 'PAY_4_6', 'PAY_4_7', 'PAY_4_8', 'PAY_5_-999',
       'PAY_5_2', 'PAY_5_3', 'PAY_5_4', 'PAY_5_5', 'PAY_5_6', 'PAY_5_7',
       'PAY_5_8', 'PAY_6_-999', 'PAY_6_2', 'PAY_6_3', 'PAY_6_4', 'PAY_6_5',
       'PAY_6_6', 'PAY_6_7', 'PAY_6_8', 'default_payment_next_month'],
      dtype='object')

In [144]:
#import pickle

# Guardar las columnas del dataset post preprocesamiento y OneHotEncoding en un archivo físico
with open('cols_preproc_OneHotEnc.pickle', 'wb') as handle:
    pickle.dump(df_defcredcard_preproc_OneHotEnc.columns, handle, protocol=pickle.HIGHEST_PROTOCOL)

In [145]:
# Importar las columnas del dataset post preprocesamiento y OneHotEncoding en un archivo físico
#with open('cols_preproc_OneHotEnc.pickle', 'rb') as handle:
#    cols_defcredcar_preproc_ohe = pickle.load(handle)

In [146]:
#cols_defcredcar_preproc_ohe

In [147]:
# Reemplazar categorías faltantes en dataset nuevo con categorias de dataset original y 
# reemplazar valores nulos con un valor definido (0)
#pd.get_dummies(df_defcredcard_preproc_OneHotEnc_test).reindex(columns = cols_defcredcar_preproc_ohe, fill_value=0)

In [148]:
# Guardar dataset post preprocesamiento y OneHotEncoding en un archivo físico
with open('df_preproc_OneHotEnc.pickle', 'wb') as f:
    pickle.dump(df_defcredcard_preproc_OneHotEnc, f)

In [149]:
# Importar dataset post preprocesamiento y OneHotEncoding en un archivo físico
with open('df_preproc_OneHotEnc.pickle', 'rb') as f:
   df_defcredcard_preproc_ohe = pickle.load(f)

In [150]:
df_defcredcard_preproc_ohe.shape

(29965, 73)

#### **REDUCCIÓN DE DIMENSIONALIDAD**

## **DIVISIÓN DE LOS DATOS**

In [151]:
X = df_defcredcard_preproc_ohe.iloc[:,:-1]
y = df_defcredcard_preproc_ohe.iloc[:,-1]

In [152]:
#from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42, stratify=y)

In [153]:
X_train.shape, X_test.shape

((23972, 72), (5993, 72))

In [154]:
y_train.shape, y_test.shape

((23972,), (5993,))

## **BALANCEO DE DATOS**

#### Balanceo de Datos (Dataset de Entrenamiento)

In [155]:
#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable Objetivo
#---------------------------------------------------------------------------------
pd.DataFrame({'Cantidad': y_train.value_counts(), '%': y_train.value_counts(normalize=True) * 100})

Cantidad          %
0     18668  77.874187
1      5304  22.125813

In [205]:
from imblearn.over_sampling import SMOTE

smote_train = SMOTE(ratio='minority')
X_smt_train, y_smt_train = smote_train.fit_sample(X_train, y_train)

columns_x = X_train.columns.to_list()
columns_y = 'default_payment_next_month'

X_bal_smt_train = pd.DataFrame(data = X_smt_train, columns = columns_x)
y_bal_smt_train = pd.Series(data = y_smt_train, name=columns_y)

In [206]:
pd.DataFrame({'Cantidad': y_bal_smt_train.value_counts(), '%': y_bal_smt_train.value_counts(normalize=True) * 100})

Cantidad     %
1     18668  50.0
0     18668  50.0

#### Balanceo de Datos (Dataset de Pruebas)

In [158]:
#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable Objetivo
#---------------------------------------------------------------------------------
pd.DataFrame({'Cantidad': y_test.value_counts(), '%': y_test.value_counts(normalize=True) * 100})

Cantidad          %
0      4667  77.874187
1      1326  22.125813

In [207]:
from imblearn.over_sampling import SMOTE

smote_valid_t = SMOTE(ratio='minority')
X_smt_test, y_smt_test = smote_valid_t.fit_sample(X_test, y_test)

columns_x = X_test.columns.to_list()
columns_y = 'default_payment_next_month'


X_bal_smt_test = pd.DataFrame(data = X_smt_test, columns = columns_x)
y_bal_smt_test = pd.Series(data = y_smt_test, name=columns_y)

In [208]:
#---------------------------------------------------------------------------------
# Tabla de Contingencia de Variable Objetivo
#---------------------------------------------------------------------------------
pd.DataFrame({'Cantidad': y_bal_smt_test.value_counts(), '%': y_bal_smt_test.value_counts(normalize=True) * 100})

Cantidad     %
1      4667  50.0
0      4667  50.0

###############################################################################
# **4. MODELADO**
###############################################################################

## **4.1 REGRESIÓN LOGÍSTICA**

### **4.1.1 Usando variables predictoras sin escalar**

In [170]:
from sklearn.linear_model import LogisticRegression

# Instanciar el modelo
model_reglog_0 = LogisticRegression()

# Entrenamiento del modelo
model_reglog_0.fit(X_train, y_train)

# Predicción de la clase
y_pred_reglog_0_train = model_reglog_0.predict(X_train)
y_pred_reglog_0_test = model_reglog_0.predict(X_test)

# Predicción de la probabilidad
y_pred_proba_reglog_0_train = model_reglog_0.predict_proba(X_train)[:,1]
y_pred_proba_reglog_0_test = model_reglog_0.predict_proba(X_test)[:,1]

In [172]:
from sklearn.metrics import accuracy_score

# Métrica Accuracy / AUC en Hold-Out
print('Accuracy Train: ', model_reglog_0.score(X_train, y_train))
print('Accuracy Test: ', accuracy_score(y_test, y_pred_reglog_0_test))
print('AUC Train: ', roc_auc_score(y_train, y_pred_reglog_0_train))
print('AUC Test: ', roc_auc_score(y_test, y_pred_reglog_0_test))

Accuracy Train:  0.7785750041715335
Accuracy Test:  0.7785750041715335
AUC Train:  0.5001628034972292
AUC Test:  0.5001628034972292


In [175]:
from sklearn.model_selection import cross_val_score
import warnings
warnings.filterwarnings('ignore')

# Métrica Accuracy / AUC en Cross Validation K-Folds
n_knnfolds = 5
print('Accuracy Prom.: ', np.mean(cross_val_score(model_reglog_0, X_train, y_train, scoring='accuracy', cv=n_knnfolds)))
print('AUC Prom.: ', np.mean(cross_val_score(model_reglog_0, X_train, y_train, scoring='roc_auc', cv=n_knnfolds)))

Accuracy Prom.:  0.7785332900049289
AUC Prom.:  0.65545972578706


In [185]:
from sklearn.model_selection import GridSearchCV

# Especificando parametros
hiperparametros = {
                    'penalty':np.array(['l1','l2']),
                    #'C':[0.001,.009,0.01,.09,1,5,10,25]
                    'C': np.logspace(-4, 4, 50)
                  }

print('Los valores a probar en la búsqueda por grilla son:')
pprint(hiperparametros)

print()
print('Si se probara todas las combinaciones se requeriría entrenar', 
      len(hiperparametros['penalty']) *
      len(hiperparametros['C']), 
      'modelos'
      )

Los valores a probar en la búsqueda por grilla son:
{'C': array([1.00000000e-04, 1.45634848e-04, 2.12095089e-04, 3.08884360e-04,
       4.49843267e-04, 6.55128557e-04, 9.54095476e-04, 1.38949549e-03,
       2.02358965e-03, 2.94705170e-03, 4.29193426e-03, 6.25055193e-03,
       9.10298178e-03, 1.32571137e-02, 1.93069773e-02, 2.81176870e-02,
       4.09491506e-02, 5.96362332e-02, 8.68511374e-02, 1.26485522e-01,
       1.84206997e-01, 2.68269580e-01, 3.90693994e-01, 5.68986603e-01,
       8.28642773e-01, 1.20679264e+00, 1.75751062e+00, 2.55954792e+00,
       3.72759372e+00, 5.42867544e+00, 7.90604321e+00, 1.15139540e+01,
       1.67683294e+01, 2.44205309e+01, 3.55648031e+01, 5.17947468e+01,
       7.54312006e+01, 1.09854114e+02, 1.59985872e+02, 2.32995181e+02,
       3.39322177e+02, 4.94171336e+02, 7.19685673e+02, 1.04811313e+03,
       1.52641797e+03, 2.22299648e+03, 3.23745754e+03, 4.71486636e+03,
       6.86648845e+03, 1.00000000e+04]),
 'penalty': array(['l1', 'l2'], dtype='<U2')}

Si

In [186]:
# Instanciar el modelo
model_reglog_grid_1 = GridSearchCV(LogisticRegression(), # modelo
                              hiperparametros, # hiperparametros
                              verbose=2, # mostrar procesamiento de GridSearchCV
                              n_jobs=-1, # paralelizar en todos los cores disponibles
                              cv=5) # numero de k-folds o grupos para validación cruzada

# Entrenamiento del modelo
model_reglog_grid_1.fit(X_train, y_train)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  70 tasks      | elapsed:   14.6s
[Parallel(n_jobs=-1)]: Done 311 tasks      | elapsed:  1.1min
[Parallel(n_jobs=-1)]: Done 500 out of 500 | elapsed:  1.8min finished


GridSearchCV(cv=5, error_score=nan,
             estimator=LogisticRegression(C=1.0, class_weight=None, dual=False,
                                          fit_intercept=True,
                                          intercept_scaling=1, l1_ratio=None,
                                          max_iter=100, multi_class='auto',
                                          n_jobs=None, penalty='l2',
                                          random_state=None, solver='lbfgs',
                                          tol=0.0001, verbose=0,
                                          warm_start=False),
             iid='deprecated', n_jobs=-1,
             param_grid={'C': array([1.00000000e-04, 1.456...
       1.67683294e+01, 2.44205309e+01, 3.55648031e+01, 5.17947468e+01,
       7.54312006e+01, 1.09854114e+02, 1.59985872e+02, 2.32995181e+02,
       3.39322177e+02, 4.94171336e+02, 7.19685673e+02, 1.04811313e+03,
       1.52641797e+03, 2.22299648e+03, 3.23745754e+03, 4.71486636e+03,
       6

In [187]:
model_reglog_grid_1_best = model_reglog_grid_1.best_estimator_
print('Mejor modelo: ')
pprint(model_reglog_grid_1_best)
print()

print('Los hiperparámetros del mejor modelo son:')
pprint(model_reglog_grid_1.best_params_)
print()

print('Accuracy luego de búsqueda en grilla en entrenamiento:', model_reglog_grid_1_best.score(X_train, y_train))
print('Accuracy luego de búsqueda en grilla en pruebas:', model_reglog_grid_1_best.score(X_test, y_test))

y_pred_proba_reglog_grid_1_best_train = model_reglog_grid_1_best.predict_proba(X_train)[::,1]
y_pred_proba_reglog_grid_1_best_test = model_reglog_grid_1_best.predict_proba(X_test)[::,1]

print(f"AUC luego de búsqueda en grilla en entrenamiento es: {roc_auc_score(y_train, y_pred_proba_reglog_grid_1_best_train)}")
print(f"AUC luego de búsqueda en grilla en pruebas es: {roc_auc_score(y_test, y_pred_proba_reglog_grid_1_best_test)}")

Mejor modelo: 
LogisticRegression(C=3.727593720314938, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

Los hiperparámetros del mejor modelo son:
{'C': 3.727593720314938, 'penalty': 'l2'}

Accuracy luego de búsqueda en grilla en entrenamiento: 0.7785750041715335
Accuracy luego de búsqueda en grilla en pruebas: 0.7789087268479893
AUC luego de búsqueda en grilla en entrenamiento es: 0.6561257259904836
AUC luego de búsqueda en grilla en pruebas es: 0.6493661894221519


In [203]:
fpr_regloggridbest1_train, tpr_regloggridbest1_train, thr_regloggridbest1_train = metrics.roc_curve(y_train, y_pred_proba_reglog_grid_1_best_train)
fpr_regloggridbest1_test, tpr_regloggridbest1_test, thr_regloggridbest1_test = metrics.roc_curve(y_test, y_pred_proba_reglog_grid_1_best_test)
roc_auc_regloggridbest1_train = auc(fpr_regloggridbest1_train, tpr_regloggridbest1_train)
roc_auc_regloggridbest1_test = auc(fpr_regloggridbest1_test, tpr_regloggridbest1_test)

print('Training AUC =', roc_auc_regloggridbest1_train)
print('Validation AUC =', roc_auc_regloggridbest1_test)

label_train = 'AUC Entrenamiento:' + ' {0:.6f}'.format(roc_auc_regloggridbest1_train)
label_test = 'AUC Pruebas:' + ' {0:.6f}'.format(roc_auc_regloggridbest1_test)
plt.plot(fpr_regloggridbest1_train, tpr_regloggridbest1_train, c = 'r', label = label_train, linewidth = 2)
plt.plot(fpr_regloggridbest1_test, tpr_regloggridbest1_test, c = 'b', label = label_test, linewidth = 2)
plt.plot([0,1], [0,1], color='orange', linestyle='--')
plt.xlabel('False Positive Rate', fontsize = 16)
plt.ylabel('True Positive Rate', fontsize = 16)
plt.title('Curva ROC / Area AUC - Regresion Logistica', fontsize = 16)
plt.legend(loc = 'lower right', fontsize = 10)
plt.show()

Training AUC = 0.6561257259904836
Validation AUC = 0.6493661894221519


### **4.1.2 Usando variable objetivo balanceada**

In [211]:
from sklearn.linear_model import LogisticRegression

# Instanciar el modelo
model_reglog_2 = LogisticRegression()

# Entrenamiento del modelo
model_reglog_2.fit(X_bal_smt_train, y_bal_smt_train)

# Predicción de la clase
y_pred_reglog_2_train = model_reglog_2.predict(X_bal_smt_train)
y_pred_reglog_2_test = model_reglog_2.predict(X_bal_smt_test)

# Predicción de la probabilidad
y_pred_proba_reglog_2_train = model_reglog_2.predict_proba(X_bal_smt_train)[:,1]
y_pred_proba_reglog_2_test = model_reglog_2.predict_proba(X_bal_smt_test)[:,1]

In [214]:
from sklearn.metrics import accuracy_score

# Métrica Accuracy / AUC en Hold-Out
print('Accuracy Train: ', model_reglog_2.score(X_bal_smt_train, y_bal_smt_train))
print('Accuracy Test: ', accuracy_score(y_bal_smt_test, y_pred_reglog_2_test))
print('AUC Train: ', roc_auc_score(y_bal_smt_train, y_pred_reglog_2_train))
print('AUC Test: ', roc_auc_score(y_bal_smt_test, y_pred_reglog_2_test))

Accuracy Train:  0.5982161988429398
Accuracy Test:  0.6002785515320335
AUC Train:  0.5982161988429398
AUC Test:  0.6002785515320335


In [215]:
from sklearn.model_selection import cross_val_score
import warnings
warnings.filterwarnings('ignore')

# Métrica Accuracy / AUC en Cross Validation K-Folds
n_knnfolds = 5
print('Accuracy Prom.: ', np.mean(cross_val_score(model_reglog_2, X_bal_smt_train, y_bal_smt_train, scoring='accuracy', cv=n_knnfolds)))
print('AUC Prom.: ', np.mean(cross_val_score(model_reglog_2, X_bal_smt_train, y_bal_smt_train, scoring='roc_auc', cv=n_knnfolds)))

Accuracy Prom.:  0.5781554246648115
AUC Prom.:  0.6547742744930999


In [216]:
from sklearn.model_selection import GridSearchCV

# Especificando parametros
hiperparametros = {
                    'penalty': np.array(['l1','l2']),
                    #'C':[0.001,.009,0.01,.09,1,5,10,25]
                    'C': np.logspace(-4, 4, 50)
                  }

print('Los valores a probar en la búsqueda por grilla son:')
pprint(hiperparametros)

print()
print('Si se probara todas las combinaciones se requeriría entrenar', 
      len(hiperparametros['penalty']) *
      len(hiperparametros['C']), 
      'modelos'
      )

Los valores a probar en la búsqueda por grilla son:
{'C': array([1.00000000e-04, 1.45634848e-04, 2.12095089e-04, 3.08884360e-04,
       4.49843267e-04, 6.55128557e-04, 9.54095476e-04, 1.38949549e-03,
       2.02358965e-03, 2.94705170e-03, 4.29193426e-03, 6.25055193e-03,
       9.10298178e-03, 1.32571137e-02, 1.93069773e-02, 2.81176870e-02,
       4.09491506e-02, 5.96362332e-02, 8.68511374e-02, 1.26485522e-01,
       1.84206997e-01, 2.68269580e-01, 3.90693994e-01, 5.68986603e-01,
       8.28642773e-01, 1.20679264e+00, 1.75751062e+00, 2.55954792e+00,
       3.72759372e+00, 5.42867544e+00, 7.90604321e+00, 1.15139540e+01,
       1.67683294e+01, 2.44205309e+01, 3.55648031e+01, 5.17947468e+01,
       7.54312006e+01, 1.09854114e+02, 1.59985872e+02, 2.32995181e+02,
       3.39322177e+02, 4.94171336e+02, 7.19685673e+02, 1.04811313e+03,
       1.52641797e+03, 2.22299648e+03, 3.23745754e+03, 4.71486636e+03,
       6.86648845e+03, 1.00000000e+04]),
 'penalty': array(['l1', 'l2'], dtype='<U2')}

Si

In [217]:
# Instanciar el modelo
model_reglog_grid_3 = GridSearchCV(LogisticRegression(), # modelo
                                   hiperparametros, # hiperparametros
                                   verbose=2, # mostrar procesamiento de GridSearchCV
                                   n_jobs=-1, # paralelizar en todos los cores disponibles
                                   cv=5) # numero de k-folds o grupos para validación cruzada

# Entrenamiento del modelo
model_reglog_grid_3.fit(X_bal_smt_train, y_bal_smt_train)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  37 tasks      | elapsed:   12.0s
[Parallel(n_jobs=-1)]: Done 158 tasks      | elapsed:   48.2s
[Parallel(n_jobs=-1)]: Done 361 tasks      | elapsed:  1.8min
[Parallel(n_jobs=-1)]: Done 500 out of 500 | elapsed:  2.5min finished


GridSearchCV(cv=5, error_score=nan,
             estimator=LogisticRegression(C=1.0, class_weight=None, dual=False,
                                          fit_intercept=True,
                                          intercept_scaling=1, l1_ratio=None,
                                          max_iter=100, multi_class='auto',
                                          n_jobs=None, penalty='l2',
                                          random_state=None, solver='lbfgs',
                                          tol=0.0001, verbose=0,
                                          warm_start=False),
             iid='deprecated', n_jobs=-1,
             param_grid={'C': array([1.00000000e-04, 1.456...
       1.67683294e+01, 2.44205309e+01, 3.55648031e+01, 5.17947468e+01,
       7.54312006e+01, 1.09854114e+02, 1.59985872e+02, 2.32995181e+02,
       3.39322177e+02, 4.94171336e+02, 7.19685673e+02, 1.04811313e+03,
       1.52641797e+03, 2.22299648e+03, 3.23745754e+03, 4.71486636e+03,
       6

In [219]:
model_reglog_grid_3_best = model_reglog_grid_3.best_estimator_
print('Mejor modelo: ')
pprint(model_reglog_grid_3_best)
print()

print('Los hiperparámetros del mejor modelo son:')
pprint(model_reglog_grid_3.best_params_)
print()

print('Accuracy luego de búsqueda en grilla en entrenamiento:', model_reglog_grid_3_best.score(X_bal_smt_train, y_bal_smt_train))
print('Accuracy luego de búsqueda en grilla en pruebas:', model_reglog_grid_3_best.score(X_bal_smt_test, y_bal_smt_test))

y_pred_proba_reglog_grid_3_best_train = model_reglog_grid_3_best.predict_proba(X_bal_smt_train)[::,1]
y_pred_proba_reglog_grid_3_best_test = model_reglog_grid_3_best.predict_proba(X_bal_smt_test)[::,1]

print(f"AUC luego de búsqueda en grilla en entrenamiento es: {roc_auc_score(y_bal_smt_train, y_pred_proba_reglog_grid_3_best_train)}")
print(f"AUC luego de búsqueda en grilla en pruebas es: {roc_auc_score(y_bal_smt_test, y_pred_proba_reglog_grid_3_best_test)}")

Mejor modelo: 
LogisticRegression(C=6866.488450042998, class_weight=None, dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

Los hiperparámetros del mejor modelo son:
{'C': 6866.488450042998, 'penalty': 'l2'}

Accuracy luego de búsqueda en grilla en entrenamiento: 0.5917345189629312
Accuracy luego de búsqueda en grilla en pruebas: 0.5955646025283908
AUC luego de búsqueda en grilla en entrenamiento es: 0.6563911085080135
AUC luego de búsqueda en grilla en pruebas es: 0.6599177609325313


In [220]:
fpr_regloggridbest3_train, tpr_regloggridbest3_train, thr_regloggridbest3_train = metrics.roc_curve(y_bal_smt_train, y_pred_proba_reglog_grid_3_best_train)
fpr_regloggridbest3_test, tpr_regloggridbest3_test, thr_regloggridbest3_test = metrics.roc_curve(y_bal_smt_test, y_pred_proba_reglog_grid_3_best_test)
roc_auc_regloggridbest3_train = auc(fpr_regloggridbest3_train, tpr_regloggridbest3_train)
roc_auc_regloggridbest3_test = auc(fpr_regloggridbest3_test, tpr_regloggridbest3_test)

print('Training AUC =', roc_auc_regloggridbest3_train)
print('Validation AUC =', roc_auc_regloggridbest3_test)

label_train = 'AUC Entrenamiento:' + ' {0:.6f}'.format(roc_auc_regloggridbest3_train)
label_test = 'AUC Pruebas:' + ' {0:.6f}'.format(roc_auc_regloggridbest3_test)
plt.plot(fpr_regloggridbest3_train, tpr_regloggridbest3_train, c = 'r', label = label_train, linewidth = 2)
plt.plot(fpr_regloggridbest3_test, tpr_regloggridbest3_test, c = 'b', label = label_test, linewidth = 2)
plt.plot([0,1], [0,1], color='orange', linestyle='--')
plt.xlabel('False Positive Rate', fontsize = 16)
plt.ylabel('True Positive Rate', fontsize = 16)
plt.title('Curva ROC / Area AUC - Regresion Logistica', fontsize = 16)
plt.legend(loc = 'lower right', fontsize = 10)
plt.show()

Training AUC = 0.6563911085080135
Validation AUC = 0.6599177609325313


## **4.2 ÁRBOL DE DECISIÓN**

In [240]:
model_tree = DecisionTreeClassifier(class_weight='balanced',random_state=0)

In [241]:
#from pprint import pprint
n_kfolds = 5
max_depth = [x for x in range(1, 40, 5)]
min_samples_leaf = [x for x in range(1, 40, 5)]
min_samples_split = [x for x in range(2, 80, 5)]
#n_estimators = [x for x in range(100, 400, 50)]
max_features = [x for x in range(1, 20, 2)]

grid = {
        'max_depth': max_depth,
        'min_samples_leaf': min_samples_leaf,
        'min_samples_split': min_samples_split,
         #'n_estimators': n_estimators,
        'max_features': max_features
        }

print('Los valores a probar en la búsqueda por grilla son:')
print(grid)

Los valores a probar en la búsqueda por grilla son:
{'max_depth': [1, 6, 11, 16, 21, 26, 31, 36], 'min_samples_leaf': [1, 6, 11, 16, 21, 26, 31, 36], 'min_samples_split': [2, 7, 12, 17, 22, 27, 32, 37, 42, 47, 52, 57, 62, 67, 72, 77], 'max_features': [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]}


In [242]:
model_dt_grid = GridSearchCV(estimator = model_tree, 
                        param_grid = grid, 
                        cv = n_kfolds,          # Nro Grupos para Validación Cruzada
                        verbose=2, 
                        n_jobs = -1      # Paralelizar en todos los cores disponibles
                        )
model_dt_grid.fit(X_bal_smt_train, y_bal_smt_train)

Fitting 5 folds for each of 10240 candidates, totalling 51200 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done 128 tasks      | elapsed:    3.1s
[Parallel(n_jobs=-1)]: Done 612 tasks      | elapsed:   13.9s
[Parallel(n_jobs=-1)]: Done 1424 tasks      | elapsed:   34.3s
[Parallel(n_jobs=-1)]: Done 2556 tasks      | elapsed:  1.1min
[Parallel(n_jobs=-1)]: Done 4016 tasks      | elapsed:  1.8min
[Parallel(n_jobs=-1)]: Done 5796 tasks      | elapsed:  2.9min
[Parallel(n_jobs=-1)]: Done 7904 tasks      | elapsed:  4.1min
[Parallel(n_jobs=-1)]: Done 10332 tasks      | elapsed:  6.4min
[Parallel(n_jobs=-1)]: Done 13088 tasks      | elapsed: 10.2min
[Parallel(n_jobs=-1)]: Done 16164 tasks      | elapsed: 13.2min
[Parallel(n_jobs=-1)]: Done 18323 tasks      | elapsed: 17.2min
[Parallel(n_jobs=-1)]: Done 19290 tasks      | elapsed: 19.2min
[Parallel(n_jobs=-1)]: Done 23186 tasks      | elapsed: 23.7min
[Parallel(n_jobs=-1)]: Done 24672 tasks      | elapsed: 26.9min
[Parallel(n_jobs=-1)]: Done 26244 tas

GridSearchCV(cv=5, error_score=nan,
             estimator=DecisionTreeClassifier(ccp_alpha=0.0,
                                              class_weight='balanced',
                                              criterion='gini', max_depth=None,
                                              max_features=None,
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              presort='deprecated',
                                              random_state=0, splitter='best'),
             iid='deprecated', n_jobs=-1,
             param_grid={'max_depth': [1, 6, 11, 16, 21, 26, 31, 36],
       

In [243]:
model_dt_grid_best = model_dt_grid.best_estimator_
print('Mejor modelo: ')
print(model_dt_grid_best)
print()

print('Los hiperparámetros del mejor modelo son:')
print(model_dt_grid.best_params_)
print()

# mejor parametros
#max_depth_best = model_dt_grid_best.max_depth

print('Accuracy luego de búsqueda en grilla en entrenamiento:', model_dt_grid_best.score(X_bal_smt_train, y_bal_smt_train))
print('Accuracy luego de búsqueda en grilla en validación:', model_dt_grid_best.score(X_bal_smt_test, y_bal_smt_test))

y_pred_proba_train_tree_best = model_dt_grid_best.predict_proba(X_bal_smt_train)[::,1]
y_pred_proba_test_tree_best = model_dt_grid_best.predict_proba(X_bal_smt_test)[::,1]

print(f"AUC luego de búsqueda en grilla en entrenamiento es: {roc_auc_score(y_bal_smt_train, y_pred_proba_train_tree_best)}")
print(f"AUC luego de búsqueda en grilla en validación es: {roc_auc_score(y_bal_smt_test, y_pred_proba_test_tree_best)}")

Mejor modelo: 
DecisionTreeClassifier(ccp_alpha=0.0, class_weight='balanced', criterion='gini',
                       max_depth=11, max_features=19, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

Los hiperparámetros del mejor modelo son:
{'max_depth': 11, 'max_features': 19, 'min_samples_leaf': 1, 'min_samples_split': 2}

Accuracy luego de búsqueda en grilla en entrenamiento: 0.8303246196700236
Accuracy luego de búsqueda en grilla en validación: 0.7611956288836512
AUC luego de búsqueda en grilla en entrenamiento es: 0.9110644671689019
AUC luego de búsqueda en grilla en validación es: 0.808387825675986


In [273]:
# CURVA AUC
fpr_train_dtbest, tpr_train_dtbest, thr_train_dtbest = metrics.roc_curve(y_bal_smt_train, y_pred_proba_train_tree_best)
fpr_valid_dtbest, tpr_valid_dtbest, thr_valid_dtbest = metrics.roc_curve(y_bal_smt_test, y_pred_proba_test_tree_best)
roc_auc_train_dtbest = auc(fpr_train_dtbest, tpr_train_dtbest)
roc_auc_valid_dtbest = auc(fpr_valid_dtbest, tpr_valid_dtbest)

print('Training AUC =', roc_auc_train_dtbest)
print('Validation AUC =', roc_auc_valid_dtbest)

label_train = 'AUC Entrenamiento:' + ' {0:.6f}'.format(roc_auc_train_dtbest)
label_valid = 'AUC Validación:' + ' {0:.6f}'.format(roc_auc_valid_dtbest)
plt.plot(fpr_train_dtbest, tpr_train_dtbest, c = 'r', label = label_train, linewidth = 2)
plt.plot(fpr_valid_dtbest, tpr_valid_dtbest, c = 'b', label = label_valid, linewidth = 2)
plt.plot([0,1], [0,1], color='orange', linestyle='--')
plt.xlabel('False Positive Rate', fontsize = 16)
plt.ylabel('True Positive Rate', fontsize = 16)
plt.title('Curva ROC / Area AUC - Arbol de Decision', fontsize = 16)
plt.legend(loc = 'lower right', fontsize = 10)

Training AUC = 0.9110644671689019
Validation AUC = 0.808387825675986


## **4.3 RANDOM FOREST**

In [245]:
from sklearn.ensemble import RandomForestClassifier
model_rndForest = RandomForestClassifier(random_state=0)

In [246]:
grid_forest = { 
    'n_estimators': [4,5,6,7,8],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth' : [4,5,6,7,8],
    'criterion' :['gini', 'entropy']
}

In [247]:
model_dt_grid_forest = GridSearchCV(estimator=model_rndForest, 
                                    param_grid=grid_forest, cv= 5,verbose=2, 
                                    n_jobs = -1)
model_dt_grid_forest.fit(X_bal_smt_train, y_bal_smt_train)

Fitting 5 folds for each of 150 candidates, totalling 750 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  37 tasks      | elapsed:    4.8s
[Parallel(n_jobs=-1)]: Done 158 tasks      | elapsed:   22.0s
[Parallel(n_jobs=-1)]: Done 361 tasks      | elapsed:   58.1s
[Parallel(n_jobs=-1)]: Done 644 tasks      | elapsed:  1.9min
[Parallel(n_jobs=-1)]: Done 750 out of 750 | elapsed:  2.3min finished


GridSearchCV(cv=5, error_score=nan,
             estimator=RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                              class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              max_samples=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators=100, n_jobs=None,
                                              oob_score=False, random_state=0,
                                   

In [248]:
model_grid_forest_best = model_dt_grid_forest.best_estimator_
print('Mejor modelo: ')
print(model_grid_forest_best)
print()

print('Los hiperparámetros del mejor modelo son:')
print(model_grid_forest_best.estimator_params)
print()

# mejor parametros
#max_depth_best = model_dt_grid_best.max_depth

print('Accuracy luego de búsqueda en grilla en entrenamiento:', model_grid_forest_best.score(X_bal_smt_train, y_bal_smt_train))
print('Accuracy luego de búsqueda en grilla en validación:', model_grid_forest_best.score(X_bal_smt_test, y_bal_smt_test))

y_pred_proba_train_forest_best = model_grid_forest_best.predict_proba(X_bal_smt_train)[::,1]
y_pred_proba_test_forest_best = model_grid_forest_best.predict_proba(X_bal_smt_test)[::,1]

print(f"AUC luego de búsqueda en grilla en entrenamiento es: {roc_auc_score(y_bal_smt_train, y_pred_proba_train_forest_best)}")
print(f"AUC luego de búsqueda en grilla en validación es: {roc_auc_score(y_bal_smt_test, y_pred_proba_test_forest_best)}")

Mejor modelo: 
RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=8, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=7,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

Los hiperparámetros del mejor modelo son:
('criterion', 'max_depth', 'min_samples_split', 'min_samples_leaf', 'min_weight_fraction_leaf', 'max_features', 'max_leaf_nodes', 'min_impurity_decrease', 'min_impurity_split', 'random_state', 'ccp_alpha')

Accuracy luego de búsqueda en grilla en entrenamiento: 0.8028176558817227
Accuracy luego de búsqueda en grilla en validación: 0.7831583458324406
AUC luego de búsqueda en grilla en entrenamiento

In [256]:
# CURVA AUC
fpr_train_rfbest, tpr_train_rfbest, thr_train_rfbest = metrics.roc_curve(y_bal_smt_train, y_pred_proba_train_forest_best)
fpr_valid_rfbest, tpr_valid_rfbest, thr_valid_rfbest = metrics.roc_curve(y_bal_smt_test, y_pred_proba_test_forest_best)
roc_auc_train_rfbest = auc(fpr_train_rfbest, tpr_train_rfbest)
roc_auc_valid_rfbest = auc(fpr_valid_rfbest, tpr_valid_rfbest)

print('Training AUC =', roc_auc_train_rfbest)
print('Validation AUC =', roc_auc_valid_rfbest)

label_train = 'AUC Entrenamiento:' + ' {0:.6f}'.format(roc_auc_train_rfbest)
label_valid = 'AUC Validación:' + ' {0:.6f}'.format(roc_auc_valid_rfbest)
plt.plot(fpr_train_rfbest, tpr_train_rfbest, c = 'r', label = label_train, linewidth = 2)
plt.plot(fpr_valid_rfbest, tpr_valid_rfbest, c = 'b', label = label_valid, linewidth = 2)
plt.plot([0,1], [0,1], color='orange', linestyle='--')
plt.xlabel('False Positive Rate', fontsize = 16)
plt.ylabel('True Positive Rate', fontsize = 16)
plt.title('Curva ROC / Area AUC - Arbol de Decision', fontsize = 16)
plt.legend(loc = 'lower right', fontsize = 10)

Training AUC = 0.8808416348386883
Validation AUC = 0.8567334648278131


## **4.4 LIGHT GBM**

### **4.4.1 Usando variables predictoras sin escalar**

In [221]:
import lightgbm as lgb

# Instanciar el modelo
model_lightgbm_0 = lgb.LGBMClassifier()

# Entrenamiento del modelo
model_lightgbm_0.fit(X_train, y_train)

# Predicción de la clase
y_pred_lightgbm0_train = model_lightgbm_0.predict(X_train)
y_pred_lightgbm0_test = model_lightgbm_0.predict(X_test)

# Predicción de la probabilidad
y_pred_proba_lightgbm0_train = model_lightgbm_0.predict_proba(X_train)[:,1]
y_pred_proba_lightgbm0_test = model_lightgbm_0.predict_proba(X_test)[:,1]

In [222]:
from sklearn.metrics import accuracy_score

# Métrica Accuracy / AUC en Hold-Out
print('Accuracy Train: ', model_lightgbm_0.score(X_train, y_train))
print('Accuracy Test: ', accuracy_score(y_test, y_pred_lightgbm0_test))
print('AUC Train: ', roc_auc_score(y_train, y_pred_lightgbm0_train))
print('AUC Test: ', roc_auc_score(y_test, y_pred_lightgbm0_test))

Accuracy Train:  0.8452778241281496
Accuracy Test:  0.8202903387285166
AUC Train:  0.6978674317380692
AUC Test:  0.6600263846700026


In [223]:
from sklearn.model_selection import cross_val_score
import warnings
warnings.filterwarnings('ignore')

# Métrica Accuracy / AUC en Cross Validation K-Folds
n_knnfolds = 5
print('Accuracy Prom.: ', np.mean(cross_val_score(model_lightgbm_0, X_train, y_train, scoring='accuracy', cv=n_knnfolds)))
print('AUC Prom.: ', np.mean(cross_val_score(model_lightgbm_0, X_train, y_train, scoring='roc_auc', cv=n_knnfolds)))

Accuracy Prom.:  0.8204989813909723
AUC Prom.:  0.7811178887792678


In [231]:
from sklearn.model_selection import GridSearchCV
from scipy.stats import randint as sp_randint
from scipy.stats import uniform as sp_uniform

# Especificando parametros
hiperparametros = {
                'learning_rate': np.logspace(-4, 4, 8),
                'max_depth': np.array([25, 50 ,100, 500, 1000]),
                'num_leaves': np.array([8, 16, 32, 64, 128]),
                'feature_fraction': np.array([0.2, 0.4, 0.6, 0.8])
                #'subsample': np.array([0.2, 0.4, 0.6, 0.8])
                  }

print('Los valores a probar en la búsqueda por grilla son:')
pprint(hiperparametros)

print()
print('Si se probara todas las combinaciones se requeriría entrenar', 
      len(hiperparametros['learning_rate']) *
      len(hiperparametros['max_depth']) *
      len(hiperparametros['num_leaves']) *
      len(hiperparametros['feature_fraction']),
      'modelos'
      )

Los valores a probar en la búsqueda por grilla son:
{'feature_fraction': array([0.2, 0.4, 0.6, 0.8]),
 'learning_rate': array([1.00000000e-04, 1.38949549e-03, 1.93069773e-02, 2.68269580e-01,
       3.72759372e+00, 5.17947468e+01, 7.19685673e+02, 1.00000000e+04]),
 'max_depth': array([  25,   50,  100,  500, 1000]),
 'num_leaves': array([  8,  16,  32,  64, 128])}

Si se probara todas las combinaciones se requeriría entrenar 800 modelos


In [232]:
# Instanciar el modelo
model_lightgbm_grid_1 = GridSearchCV(lgb.LGBMClassifier(), # modelo
                                   hiperparametros, # hiperparametros
                                   verbose=2, # mostrar procesamiento de GridSearchCV
                                   n_jobs=-1, # paralelizar en todos los cores disponibles
                                   cv=5) # numero de k-folds o grupos para validación cruzada

# Entrenamiento del modelo
model_lightgbm_grid_1.fit(X_train, y_train)

Fitting 5 folds for each of 800 candidates, totalling 4000 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  37 tasks      | elapsed:   13.7s
[Parallel(n_jobs=-1)]: Done 158 tasks      | elapsed:   55.8s
[Parallel(n_jobs=-1)]: Done 361 tasks      | elapsed:  2.1min
[Parallel(n_jobs=-1)]: Done 644 tasks      | elapsed:  3.2min
[Parallel(n_jobs=-1)]: Done 1009 tasks      | elapsed:  4.2min
[Parallel(n_jobs=-1)]: Done 1454 tasks      | elapsed:  7.6min
[Parallel(n_jobs=-1)]: Done 1981 tasks      | elapsed:  9.3min
[Parallel(n_jobs=-1)]: Done 2588 tasks      | elapsed: 14.2min
[Parallel(n_jobs=-1)]: Done 3277 tasks      | elapsed: 18.6min
[Parallel(n_jobs=-1)]: Done 4000 out of 4000 | elapsed: 22.6min finished


GridSearchCV(cv=5, error_score=nan,
             estimator=LGBMClassifier(boosting_type='gbdt', class_weight=None,
                                      colsample_bytree=1.0,
                                      importance_type='split',
                                      learning_rate=0.1, max_depth=-1,
                                      min_child_samples=20,
                                      min_child_weight=0.001,
                                      min_split_gain=0.0, n_estimators=100,
                                      n_jobs=-1, num_leaves=31, objective=None,
                                      random_state=None, reg_alpha=0.0,
                                      reg_lambda=0.0, silent=Tru...
             param_grid={'feature_fraction': array([0.2, 0.4, 0.6, 0.8]),
                         'learning_rate': array([1.00000000e-04, 1.38949549e-03, 1.93069773e-02, 2.68269580e-01,
       3.72759372e+00, 5.17947468e+01, 7.19685673e+02, 1.00000000e+04]),
             

In [260]:
model_lightgbm_grid_1_best = model_lightgbm_grid_1.best_estimator_
print('Mejor modelo: ')
pprint(model_lightgbm_grid_1_best)
print()

print('Los hiperparámetros del mejor modelo son:')
pprint(model_lightgbm_grid_1.best_params_)
print()

print('Accuracy luego de búsqueda en grilla en entrenamiento:', model_lightgbm_grid_1_best.score(X_train, y_train))
print('Accuracy luego de búsqueda en grilla en pruebas:', model_lightgbm_grid_1_best.score(X_test, y_test))

y_pred_proba_lighthbm_grid_1_best_train = model_lightgbm_grid_1_best.predict_proba(X_train)[::,1]
y_pred_proba_lighthbm_grid_1_best_test = model_lightgbm_grid_1_best.predict_proba(X_test)[::,1]

print(f"AUC luego de búsqueda en grilla en entrenamiento es: {roc_auc_score(y_train, y_pred_proba_lighthbm_grid_1_best_train)}")
print(f"AUC luego de búsqueda en grilla en pruebas es: {roc_auc_score(y_test, y_pred_proba_lighthbm_grid_1_best_test)}")

Mejor modelo: 
LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               feature_fraction=0.8, importance_type='split',
               learning_rate=0.019306977288832496, max_depth=25,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=64, objective=None,
               random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

Los hiperparámetros del mejor modelo son:
{'feature_fraction': 0.8,
 'learning_rate': 0.019306977288832496,
 'max_depth': 25,
 'num_leaves': 64}

Accuracy luego de búsqueda en grilla en entrenamiento: 0.8295928583347238
Accuracy luego de búsqueda en grilla en pruebas: 0.8167862506257301
AUC luego de búsqueda en grilla en entrenamiento es: 0.8478253088580292
AUC luego de búsqueda en grilla en pruebas es: 0.7754042940048562


In [262]:
fpr_lightgbmgridbest1_train, tpr_lightgbmgridbest1_train, thr_lightgbmgridbest1_train = metrics.roc_curve(y_train, y_pred_proba_lighthbm_grid_1_best_train)
fpr_lightgbmgridbest1_test, tpr_lightgbmgridbest1_test, thr_lightgbmgridbest1_test = metrics.roc_curve(y_test, y_pred_proba_lighthbm_grid_1_best_test)
roc_auc_lightgbmgridbest1_train = auc(fpr_lightgbmgridbest1_train, tpr_lightgbmgridbest1_train)
roc_auc_lightgbmgridbest1_test = auc(fpr_lightgbmgridbest1_test, tpr_lightgbmgridbest1_test)

print('Training AUC =', roc_auc_lightgbmgridbest1_train)
print('Validation AUC =', roc_auc_lightgbmgridbest1_test)

label_train = 'AUC Entrenamiento:' + ' {0:.6f}'.format(roc_auc_lightgbmgridbest1_train)
label_test = 'AUC Pruebas:' + ' {0:.6f}'.format(roc_auc_lightgbmgridbest1_test)
plt.plot(fpr_lightgbmgridbest1_train, tpr_lightgbmgridbest1_train, c = 'r', label = label_train, linewidth = 2)
plt.plot(fpr_lightgbmgridbest1_test, tpr_lightgbmgridbest1_test, c = 'b', label = label_test, linewidth = 2)
plt.plot([0,1], [0,1], color='orange', linestyle='--')
plt.xlabel('False Positive Rate', fontsize = 16)
plt.ylabel('True Positive Rate', fontsize = 16)
plt.title('Curva ROC / Area AUC - Regresion Logistica', fontsize = 16)
plt.legend(loc = 'lower right', fontsize = 10)
plt.show()

Training AUC = 0.8478253088580292
Validation AUC = 0.7754042940048562


### **4.4.2 Usando variable objetivo balanceada**

In [234]:
import lightgbm as lgb

# Instanciar el modelo
model_lightgbm_2 = lgb.LGBMClassifier()

# Entrenamiento del modelo
model_lightgbm_2.fit(X_bal_smt_train, y_bal_smt_train)

# Predicción de la clase
y_pred_lightgbm_2_train = model_lightgbm_2.predict(X_bal_smt_train)
y_pred_lightgbm_2_test = model_lightgbm_2.predict(X_bal_smt_test)

# Predicción de la probabilidad
y_pred_proba_lightgbm_2_train = model_lightgbm_2.predict_proba(X_bal_smt_train)[:,1]
y_pred_proba_lightgbm_2_test = model_lightgbm_2.predict_proba(X_bal_smt_test)[:,1]

In [235]:
from sklearn.metrics import accuracy_score

# Métrica Accuracy / AUC en Hold-Out
print('Accuracy Train: ', model_lightgbm_2.score(X_bal_smt_train, y_bal_smt_train))
print('Accuracy Test: ', accuracy_score(y_bal_smt_test, y_pred_lightgbm_2_test))
print('AUC Train: ', roc_auc_score(y_bal_smt_train, y_pred_lightgbm_2_train))
print('AUC Test: ', roc_auc_score(y_bal_smt_test, y_pred_lightgbm_2_test))

Accuracy Train:  0.8536265266766659
Accuracy Test:  0.8144418255838869
AUC Train:  0.853626526676666
AUC Test:  0.8144418255838869


In [236]:
from sklearn.model_selection import cross_val_score
import warnings
warnings.filterwarnings('ignore')

# Métrica Accuracy / AUC en Cross Validation K-Folds
n_knnfolds = 5
print('Accuracy Prom.: ', np.mean(cross_val_score(model_lightgbm_2, X_bal_smt_train, y_bal_smt_train, scoring='accuracy', cv=n_knnfolds)))
print('AUC Prom.: ', np.mean(cross_val_score(model_lightgbm_2, X_bal_smt_train, y_bal_smt_train, scoring='roc_auc', cv=n_knnfolds)))

Accuracy Prom.:  0.8175002397623279
AUC Prom.:  0.9007268827123529


In [237]:
from sklearn.model_selection import GridSearchCV
from scipy.stats import randint as sp_randint
from scipy.stats import uniform as sp_uniform

# Especificando parametros
hiperparametros = {
                'learning_rate': np.logspace(-4, 4, 8),
                'max_depth': np.array([25, 50 ,100, 500, 1000]),
                'num_leaves': np.array([8, 16, 32, 64, 128]),
                'feature_fraction': np.array([0.2, 0.4, 0.6, 0.8])
                #'subsample': np.array([0.2, 0.4, 0.6, 0.8])
                  }

print('Los valores a probar en la búsqueda por grilla son:')
pprint(hiperparametros)

print()
print('Si se probara todas las combinaciones se requeriría entrenar', 
      len(hiperparametros['learning_rate']) *
      len(hiperparametros['max_depth']) *
      len(hiperparametros['num_leaves']) *
      len(hiperparametros['feature_fraction']),
      'modelos'
      )

Los valores a probar en la búsqueda por grilla son:
{'feature_fraction': array([0.2, 0.4, 0.6, 0.8]),
 'learning_rate': array([1.00000000e-04, 1.38949549e-03, 1.93069773e-02, 2.68269580e-01,
       3.72759372e+00, 5.17947468e+01, 7.19685673e+02, 1.00000000e+04]),
 'max_depth': array([  25,   50,  100,  500, 1000]),
 'num_leaves': array([  8,  16,  32,  64, 128])}

Si se probara todas las combinaciones se requeriría entrenar 800 modelos


In [238]:
# Instanciar el modelo
model_lightgbm_grid_3 = GridSearchCV(lgb.LGBMClassifier(), # modelo
                                   hiperparametros, # hiperparametros
                                   verbose=2, # mostrar procesamiento de GridSearchCV
                                   n_jobs=-1, # paralelizar en todos los cores disponibles
                                   cv=5) # numero de k-folds o grupos para validación cruzada

# Entrenamiento del modelo
model_lightgbm_grid_3.fit(X_bal_smt_train, y_bal_smt_train)

Fitting 5 folds for each of 800 candidates, totalling 4000 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  37 tasks      | elapsed:   18.5s
[Parallel(n_jobs=-1)]: Done 158 tasks      | elapsed:  1.4min
[Parallel(n_jobs=-1)]: Done 361 tasks      | elapsed:  3.1min
[Parallel(n_jobs=-1)]: Done 644 tasks      | elapsed:  4.8min
[Parallel(n_jobs=-1)]: Done 1009 tasks      | elapsed:  6.3min
[Parallel(n_jobs=-1)]: Done 1454 tasks      | elapsed: 11.4min
[Parallel(n_jobs=-1)]: Done 1981 tasks      | elapsed: 14.0min
[Parallel(n_jobs=-1)]: Done 2588 tasks      | elapsed: 21.3min
[Parallel(n_jobs=-1)]: Done 3277 tasks      | elapsed: 27.9min
[Parallel(n_jobs=-1)]: Done 4000 out of 4000 | elapsed: 33.7min finished


GridSearchCV(cv=5, error_score=nan,
             estimator=LGBMClassifier(boosting_type='gbdt', class_weight=None,
                                      colsample_bytree=1.0,
                                      importance_type='split',
                                      learning_rate=0.1, max_depth=-1,
                                      min_child_samples=20,
                                      min_child_weight=0.001,
                                      min_split_gain=0.0, n_estimators=100,
                                      n_jobs=-1, num_leaves=31, objective=None,
                                      random_state=None, reg_alpha=0.0,
                                      reg_lambda=0.0, silent=Tru...
             param_grid={'feature_fraction': array([0.2, 0.4, 0.6, 0.8]),
                         'learning_rate': array([1.00000000e-04, 1.38949549e-03, 1.93069773e-02, 2.68269580e-01,
       3.72759372e+00, 5.17947468e+01, 7.19685673e+02, 1.00000000e+04]),
             

In [239]:
model_lightgbm_grid_3_best = model_lightgbm_grid_3.best_estimator_
print('Mejor modelo: ')
pprint(model_lightgbm_grid_3_best)
print()

print('Los hiperparámetros del mejor modelo son:')
pprint(model_lightgbm_grid_3.best_params_)
print()

print('Accuracy luego de búsqueda en grilla en entrenamiento:', model_lightgbm_grid_3_best.score(X_bal_smt_train, y_bal_smt_train))
print('Accuracy luego de búsqueda en grilla en pruebas:', model_lightgbm_grid_3_best.score(X_bal_smt_test, y_bal_smt_test))

y_pred_proba_lighthbm_grid_3_best_train = model_lightgbm_grid_3_best.predict_proba(X_bal_smt_train)[::,1]
y_pred_proba_lighthbm_grid_3_best_test = model_lightgbm_grid_3_best.predict_proba(X_bal_smt_test)[::,1]

print(f"AUC luego de búsqueda en grilla en entrenamiento es: {roc_auc_score(y_bal_smt_train, y_pred_proba_lighthbm_grid_3_best_train)}")
print(f"AUC luego de búsqueda en grilla en pruebas es: {roc_auc_score(y_bal_smt_test, y_pred_proba_lighthbm_grid_3_best_test)}")

Mejor modelo: 
LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               feature_fraction=0.6, importance_type='split',
               learning_rate=0.2682695795279725, max_depth=50,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=128, objective=None,
               random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

Los hiperparámetros del mejor modelo son:
{'feature_fraction': 0.6,
 'learning_rate': 0.2682695795279725,
 'max_depth': 50,
 'num_leaves': 128}

Accuracy luego de búsqueda en grilla en entrenamiento: 0.9770998500107135
Accuracy luego de búsqueda en grilla en pruebas: 0.8071566316691665
AUC luego de búsqueda en grilla en entrenamiento es: 0.9977406684364445
AUC luego de búsqueda en grilla en pruebas es: 0.8795075122966745


In [263]:
fpr_lightgbmgridbest3_train, tpr_lightgbmgridbest3_train, thr_lightgbmgridbest3_train = metrics.roc_curve(y_bal_smt_train, y_pred_proba_lighthbm_grid_3_best_train)
fpr_lightgbmgridbest3_test, tpr_lightgbmgridbest3_test, thr_lightgbmgridbest3_test = metrics.roc_curve(y_bal_smt_test, y_pred_proba_lighthbm_grid_3_best_test)
roc_auc_lightgbmgridbest3_train = auc(fpr_lightgbmgridbest3_train, tpr_lightgbmgridbest3_train)
roc_auc_lightgbmgridbest3_test = auc(fpr_lightgbmgridbest3_test, tpr_lightgbmgridbest3_test)

print('Training AUC =', roc_auc_lightgbmgridbest3_train)
print('Validation AUC =', roc_auc_lightgbmgridbest3_test)

label_train = 'AUC Entrenamiento:' + ' {0:.6f}'.format(roc_auc_lightgbmgridbest3_train)
label_test = 'AUC Pruebas:' + ' {0:.6f}'.format(roc_auc_lightgbmgridbest3_test)
plt.plot(fpr_lightgbmgridbest3_train, tpr_lightgbmgridbest3_train, c = 'r', label = label_train, linewidth = 2)
plt.plot(fpr_lightgbmgridbest3_test, tpr_lightgbmgridbest3_test, c = 'b', label = label_test, linewidth = 2)
plt.plot([0,1], [0,1], color='orange', linestyle='--')
plt.xlabel('False Positive Rate', fontsize = 16)
plt.ylabel('True Positive Rate', fontsize = 16)
plt.title('Curva ROC / Area AUC - Regresion Logistica', fontsize = 16)
plt.legend(loc = 'lower right', fontsize = 10)
plt.show()

Training AUC = 0.9977406684364445
Validation AUC = 0.8795075122966745


###############################################################################
# **5. EVALUACIÓN**
###############################################################################

In [274]:
del result_table

In [275]:
result_table = pd.DataFrame(columns=['classifiers', 'fpr','tpr','auc'])

In [276]:
result_table = result_table.append({'classifiers': model_reglog_grid_3_best.__class__.__name__,
                                    'fpr': fpr_regloggridbest3_test, 
                                    'tpr': tpr_regloggridbest3_test, 
                                    'auc': roc_auc_regloggridbest3_test}, ignore_index=True)

In [277]:
result_table = result_table.append({'classifiers': model_dt_grid_best.__class__.__name__,
                                    'fpr': fpr_valid_dtbest, 
                                    'tpr': tpr_valid_dtbest, 
                                    'auc': roc_auc_valid_dtbest}, ignore_index=True)

In [278]:
result_table = result_table.append({'classifiers': model_grid_forest_best.__class__.__name__,
                                    'fpr': fpr_valid_rfbest, 
                                    'tpr': tpr_valid_rfbest, 
                                    'auc': roc_auc_valid_rfbest}, ignore_index=True)

In [279]:
result_table = result_table.append({'classifiers': model_lightgbm_grid_3_best.__class__.__name__,
                                    'fpr': fpr_lightgbmgridbest3_test, 
                                    'tpr': tpr_lightgbmgridbest3_test, 
                                    'auc': roc_auc_lightgbmgridbest3_test}, ignore_index=True)

In [280]:
result_table

classifiers  ...       auc
0      LogisticRegression  ...  0.659918
1  DecisionTreeClassifier  ...  0.808388
2  RandomForestClassifier  ...  0.856733
3          LGBMClassifier  ...  0.879508

[4 rows x 4 columns]

In [281]:
result_table.set_index('classifiers', inplace=True)
result_table

fpr  ...       auc
classifiers                                                                ...          
LogisticRegression      [0.0, 0.00021427040925648167, 0.00021427040925...  ...  0.659918
DecisionTreeClassifier  [0.0, 0.020998500107135203, 0.0231412041997000...  ...  0.808388
RandomForestClassifier  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  ...  0.856733
LGBMClassifier          [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  ...  0.879508

[4 rows x 3 columns]

In [1]:
fig = plt.figure(figsize=(10,12))

for i in result_table.index:
    plt.plot(result_table.loc[i]['fpr'], 
             result_table.loc[i]['tpr'], 
             label="{}, AUC={:.3f}".format(i, result_table.loc[i]['auc']))
    
plt.plot([0,1], [0,1], color='orange', linestyle='--')

plt.xticks(np.arange(0.0, 1.1, step=0.1))
plt.xlabel("Flase Positive Rate", fontsize=15)

plt.yticks(np.arange(0.0, 1.1, step=0.1))
plt.ylabel("True Positive Rate", fontsize=15)

plt.title('ROC Curve Analysis', fontweight='bold', fontsize=15)
plt.legend(prop={'size':13}, loc='lower right')

plt.show()